<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>BigWig · GenomicFeatures.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>GenomicFeatures.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><a class="toctext" href="../../intervals/">Intervals</a></li><li><span class="toctext">I/O</span><ul><li><a class="toctext" href="../bed/">BED</a></li><li><a class="toctext" href="../gff3/">GFF3</a></li><li class="current"><a class="toctext" href>BigWig</a><ul class="internal"><li><a class="toctext" href="#Description-1">Description</a></li><li><a class="toctext" href="#Examples-1">Examples</a></li><li><a class="toctext" href="#API-1">API</a></li></ul></li><li><a class="toctext" href="../bigbed/">BigBed</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>I/O</li><li><a href>BigWig</a></li></ul></nav><hr/><div id="topbar"><span>BigWig</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="bigWig-1" href="#bigWig-1">bigWig</a></h1><h2><a class="nav-anchor" id="Description-1" href="#Description-1">Description</a></h2><p>bigWig is a binary file format for associating a floating point number with each base in the genome. bigWig files are indexed to quickly fetch specific regions.</p><p>I/O tools for bigWig are provided from the <code>GenomicFeatures.BigWig</code> module, which exports following three types:</p><ul><li>Reader type: <code>BigWig.Reader</code></li><li>Writer type: <code>BigWig.Writer</code></li><li>Element type: <code>BigWig.Record</code></li></ul><h2><a class="nav-anchor" id="Examples-1" href="#Examples-1">Examples</a></h2><p>A common workflow is to open a file, iterate over records, and close the file:</p><pre><code class="language-julia"># Import the BigWig module.
using GenomicFeatures

# Open a bigWig file (e.g. mapping depth or coverage).
reader = open(BigWig.Reader, &quot;data.cov.bw&quot;)

# Iterate over records overlapping with a query interval.
for record in eachoverlap(reader, Interval(&quot;Chr2&quot;, 5001, 6000))
    # Extract the start position, end position and value of the record,
    startpos = BigWig.chromstart(record)
    endpos = BigWig.chromend(record)
    value = BigWig.value(record)
    # and do something...
end

# Finally, close the reader.
close(reader)</code></pre><p><code>BigWig.values</code> is a handy function that returns a vector of values. This returns a value per position within the query region:</p><pre><code class="language-julia"># Get values in Chr2:5001-6000 as a vector of 1000 elements.
BigWig.values(reader, Interval(&quot;Chr2&quot;, 5001, 6000))</code></pre><p>Iterating over all records is also supported:</p><pre><code class="language-julia">reader = open(BigWig.Reader, &quot;data.cov.bw&quot;)
for record in reader
    # ...
end
close(reader)</code></pre><p>Creating a bigWig can be written as follows:</p><pre><code class="language-julia"># Open an output file.
file = open(&quot;data.cov.bw&quot;, &quot;w&quot;)

# Initialize a bigWig writer.
writer = BigWig.Writer(file, [(&quot;chr1&quot;, 2000), (&quot;chr2&quot;, 1000)])

# Write records.
write(writer, (&quot;chr1&quot;,   1, 100, 1.0))
write(writer, (&quot;chr1&quot;, 101, 200, 2.1))
# ...
write(writer, (&quot;chr2&quot;,  51, 150, 3.2))

# Close the writer (this closes the file, too).
close(writer)</code></pre><h2><a class="nav-anchor" id="API-1" href="#API-1">API</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BigWig.Reader" href="#GenomicFeatures.BigWig.Reader"><code>GenomicFeatures.BigWig.Reader</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">BigWig.Reader(input::IO)</code></pre><p>Create a reader for bigWig file format.</p><p>Note that <code>input</code> must be seekable.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BigWig.chromlist" href="#GenomicFeatures.BigWig.chromlist"><code>GenomicFeatures.BigWig.chromlist</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">chromlist(reader::BigWig.Reader)::Vector{Tuple{String,Int}}</code></pre><p>Get the <code>(name, length)</code> pairs of chromosomes/contigs.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BigWig.values" href="#GenomicFeatures.BigWig.values"><code>GenomicFeatures.BigWig.values</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">values(reader::BigWig.Reader, interval::Interval)::Vector{Float32}</code></pre><p>Get a vector of values within <code>interval</code> from <code>reader</code>.</p><p>This function fills missing values with <code>NaN32</code>.</p></div></div><div><div><pre><code class="language-none">values(reader::BigWig.Reader, chrom::AbstractString, range::UnitRange)::Vector{Float32}</code></pre><p>Get a vector of values within <code>range</code> of <code>chrom</code> from <code>reader</code>.</p><p>This function fills missing values with <code>NaN32</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BigWig.Writer" href="#GenomicFeatures.BigWig.Writer"><code>GenomicFeatures.BigWig.Writer</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">BigWig.Writer(output::IO, chromlist; binsize=64, datatype=:bedgraph)</code></pre><p>Create a data writer of the bigWig file format.</p><p><strong>Arguments</strong></p><ul><li><code>output</code>: data sink</li><li><code>chromlist</code>: chromosome list with length</li><li><code>binsize=64</code>: size of a zoom with the highest resolution</li><li><code>datatype=:bedgraph</code>: encoding of values (<code>:bedgraph</code>, <code>:varstep</code> or <code>:fixedstep</code>)</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia">output = open(&quot;data.bw&quot;, &quot;w&quot;)
writer = BigWig.Writer(output, [(&quot;chr1&quot;, 12345), (&quot;chr2&quot;, 9100)])
write(writer, (&quot;chr1&quot;, 501, 600, 1.0))
write(writer, (&quot;chr2&quot;, 301, 450, 3.0))
close(writer)</code></pre></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BigWig.Record" href="#GenomicFeatures.BigWig.Record"><code>GenomicFeatures.BigWig.Record</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">BigWig.Record()</code></pre><p>Create an unfilled bigWig record.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BigWig.chrom" href="#GenomicFeatures.BigWig.chrom"><code>GenomicFeatures.BigWig.chrom</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">chrom(record::Record)::String</code></pre><p>Get the chromosome name of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BigWig.chromid" href="#GenomicFeatures.BigWig.chromid"><code>GenomicFeatures.BigWig.chromid</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">chromid(record::Record)::UInt32</code></pre><p>Get the chromosome ID of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BigWig.chromstart" href="#GenomicFeatures.BigWig.chromstart"><code>GenomicFeatures.BigWig.chromstart</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">chromstart(record::Record)::Int</code></pre><p>Get the start position of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BigWig.chromend" href="#GenomicFeatures.BigWig.chromend"><code>GenomicFeatures.BigWig.chromend</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">chromend(record::Record)::Int</code></pre><p>Get the end position of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BigWig.value" href="#GenomicFeatures.BigWig.value"><code>GenomicFeatures.BigWig.value</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">value(record::Record)::Float32</code></pre><p>Get the value of <code>record</code>.</p></div></div></section><footer><hr/><a class="previous" href="../gff3/"><span class="direction">Previous</span><span class="title">GFF3</span></a><a class="next" href="../bigbed/"><span class="direction">Next</span><span class="title">BigBed</span></a></footer></article></body></html>
