<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>GFF3 · GenomicFeatures.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>GenomicFeatures.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><a class="toctext" href="../../intervals/">Intervals</a></li><li><span class="toctext">I/O</span><ul><li><a class="toctext" href="../bed/">BED</a></li><li class="current"><a class="toctext" href>GFF3</a><ul class="internal"><li><a class="toctext" href="#Description-1">Description</a></li><li><a class="toctext" href="#Examples-1">Examples</a></li><li><a class="toctext" href="#API-1">API</a></li></ul></li><li><a class="toctext" href="../bigwig/">BigWig</a></li><li><a class="toctext" href="../bigbed/">BigBed</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>I/O</li><li><a href>GFF3</a></li></ul></nav><hr/><div id="topbar"><span>GFF3</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="GFF3-1" href="#GFF3-1">GFF3</a></h1><h2><a class="nav-anchor" id="Description-1" href="#Description-1">Description</a></h2><p>GFF3 is a text-based file format for representing genomic annotations. The major difference from BED is that GFF3 is more structured and can include sequences in the FASTA file format.</p><p>I/O tools for GFF3 are provided from the <code>GenomicFeatures.GFF3</code> module, which exports following three types:</p><ul><li>Reader type: <code>GFF3.Reader</code></li><li>Writer type: <code>GFF3.Writer</code></li><li>Element type: <code>GFF3.Record</code></li></ul><p>A GFF3 file may contain directives and/or comments in addition to genomic features. These lines are skipped by default but you can control the behavior by passing keyword arguments to <code>GFF3.Reader</code>. See the docstring for details.</p><h2><a class="nav-anchor" id="Examples-1" href="#Examples-1">Examples</a></h2><p>Here is a common workflow to iterate over all records in a GFF3 file:</p><pre><code class="language-julia"># Import the GFF3 module.
using GenomicFeatures

# Open a GFF3 file.
reader = open(GFF3.Reader, &quot;data.gff3&quot;)

# Iterate over records.
for record in reader
    # Do something on record (see Accessors section).
    seqid = GFF3.seqid(reader)
    # ...
end

# Finally, close the reader
close(reader)</code></pre><p>If you are interested in directives (which starts with &#39;#&#39;) in addition to genomic features, you need to pass <code>skip_directives=false</code> when initializing a GFF3 constructor:</p><pre><code class="language-julia"># Set skip_directives to true (this is set to false by default).
reader = GFF3.Reader(open(&quot;data.gff3&quot;), skip_directives=false)
for record in record
    # Branch by record type.
    if GFF3.isfeature(record)
        # ...
    elseif GFF3.isdirective(record)
        # ...
    else
        # This never happens.
        assert(false)
    end
end
close(reader)</code></pre><p>GenomicFeatures.jl supports <a href="http://www.htslib.org/doc/tabix.html">tabix</a> to retrieve records overlapping with a specific interval. First you need to create a block compression file from a GFF3 file using bgzip and then index it using tabix.</p><pre><code class="language-none">cat data.gff3 | grep -v &quot;^#&quot; | sort -k1,1 -k4,4n | bgzip &gt;data.gff3.bgz
tabix data.gff3.bgz  # this creates data.gff3.bgz.tbi</code></pre><p>Then you can read the block compression file as follows:</p><pre><code class="language-julia"># Read the block compression gzip file.
reader = GFF3.Reader(&quot;data.gff3.bgz&quot;)
for record in eachoverlap(reader, Interval(&quot;chr1&quot;, 250_000, 300_000))
    # Each record overlap the query interval.
    # ...
end</code></pre><h2><a class="nav-anchor" id="API-1" href="#API-1">API</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.Reader" href="#GenomicFeatures.GFF3.Reader"><code>GenomicFeatures.GFF3.Reader</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">GFF3.Reader(input::IO;
            index=nothing,
            save_directives::Bool=false,
            skip_features::Bool=false,
            skip_directives::Bool=true,
            skip_comments::Bool=true)

GFF3.Reader(input::AbstractString;
            index=:auto,
            save_directives::Bool=false,
            skip_features::Bool=false,
            skip_directives::Bool=true,
            skip_comments::Bool=true)</code></pre><p>Create a reader for data in GFF3 format.</p><p>The first argument specifies the data source. When it is a filepath that ends with <em>.bgz</em>, it is considered to be block compression file format (BGZF) and the function will try to find a tabix index file (&lt;filename&gt;.tbi) and read it if any. See <a href="http://www.htslib.org/doc/tabix.html">http://www.htslib.org/doc/tabix.html</a> for bgzip and tabix tools.</p><p><strong>Arguments</strong></p><ul><li><code>input</code>: data source (<code>IO</code> object or filepath)</li><li><code>index</code>: path to a tabix file</li><li><code>save_directives</code>: flag to save directive records (which can be accessed with <code>GFF3.directives</code>)</li><li><code>skip_features</code>: flag to skip feature records</li><li><code>skip_directives</code>: flag to skip directive records</li><li><code>skip_comments</code>:  flag to skip comment records</li></ul></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.directives" href="#GenomicFeatures.GFF3.directives"><code>GenomicFeatures.GFF3.directives</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Return all directives that preceded the last GFF entry parsed as an array of strings.</p><p>Directives at the end of the file can be accessed by calling <code>close(reader)</code> and then <code>directives(reader)</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.hasfasta" href="#GenomicFeatures.GFF3.hasfasta"><code>GenomicFeatures.GFF3.hasfasta</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Return true if the GFF3 stream is at its end and there is trailing FASTA data.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.getfasta" href="#GenomicFeatures.GFF3.getfasta"><code>GenomicFeatures.GFF3.getfasta</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Return a BioSequences.FASTA.Reader initialized to parse trailing FASTA data.</p><p>Throws an exception if there is no trailing FASTA, which can be checked using <code>hasfasta</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.Writer" href="#GenomicFeatures.GFF3.Writer"><code>GenomicFeatures.GFF3.Writer</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">GFF3.Writer(output::IO)</code></pre><p>Create a data writer of the GFF3 file format.</p><p><strong>Arguments:</strong></p><ul><li><code>output</code>: data sink</li></ul></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.Record" href="#GenomicFeatures.GFF3.Record"><code>GenomicFeatures.GFF3.Record</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">GFF3.Record()</code></pre><p>Create an unfilled GFF3 record.</p></div></div><div><div><pre><code class="language-none">GFF3.Record(data::Vector{UInt8})</code></pre><p>Create a GFF3 record object from <code>data</code>. This function verifies and indexes fields for accessors. Note that the ownership of <code>data</code> is transferred to a new record object.</p></div></div><div><div><pre><code class="language-none">GFF3.Record(str::AbstractString)</code></pre><p>Create a GFF3 record object from <code>str</code>. This function verifies and indexes fields for accessors.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.isfeature" href="#GenomicFeatures.GFF3.isfeature"><code>GenomicFeatures.GFF3.isfeature</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">isfeature(record::Record)::Bool</code></pre><p>Test if <code>record</code> is a feature record.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.isdirective" href="#GenomicFeatures.GFF3.isdirective"><code>GenomicFeatures.GFF3.isdirective</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">isdirective(record::Record)::Bool</code></pre><p>Test if <code>record</code> is a directive record.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.iscomment" href="#GenomicFeatures.GFF3.iscomment"><code>GenomicFeatures.GFF3.iscomment</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">iscomment(record::Record)::Bool</code></pre><p>Test if <code>record</code> is a comment record.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.seqid" href="#GenomicFeatures.GFF3.seqid"><code>GenomicFeatures.GFF3.seqid</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">seqid(record::Record)::String</code></pre><p>Get the sequence ID of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.source" href="#GenomicFeatures.GFF3.source"><code>GenomicFeatures.GFF3.source</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">source(record::Record)::String</code></pre><p>Get the source of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.featuretype" href="#GenomicFeatures.GFF3.featuretype"><code>GenomicFeatures.GFF3.featuretype</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">featuretype(record::Record)::String</code></pre><p>Get the type of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.seqstart" href="#GenomicFeatures.GFF3.seqstart"><code>GenomicFeatures.GFF3.seqstart</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">seqstart(record::Record)::Int</code></pre><p>Get the start coordinate of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.seqend" href="#GenomicFeatures.GFF3.seqend"><code>GenomicFeatures.GFF3.seqend</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">seqend(record::Record)::Int</code></pre><p>Get the end coordinate of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.score" href="#GenomicFeatures.GFF3.score"><code>GenomicFeatures.GFF3.score</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">score(record::Record)::Float64</code></pre><p>Get the score of <code>record</code></p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.strand" href="#GenomicFeatures.GFF3.strand"><code>GenomicFeatures.GFF3.strand</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">strand(record::Record)::GenomicFeatures.Strand</code></pre><p>Get the strand of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.phase" href="#GenomicFeatures.GFF3.phase"><code>GenomicFeatures.GFF3.phase</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">phase(record::Record)::Int</code></pre><p>Get the phase of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.attributes" href="#GenomicFeatures.GFF3.attributes"><code>GenomicFeatures.GFF3.attributes</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">attributes(record::Record)::Vector{Pair{String,Vector{String}}}</code></pre><p>Get the attributes of <code>record</code>.</p></div></div><div><div><pre><code class="language-none">attributes(record::Record, key::String)::Vector{String}</code></pre><p>Get the attributes of <code>record</code> with <code>key</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.GFF3.content" href="#GenomicFeatures.GFF3.content"><code>GenomicFeatures.GFF3.content</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">content(record::Record)::String</code></pre><p>Get the content of <code>record</code>. Leading &#39;#&#39; letters are removed.</p></div></div></section><footer><hr/><a class="previous" href="../bed/"><span class="direction">Previous</span><span class="title">BED</span></a><a class="next" href="../bigwig/"><span class="direction">Next</span><span class="title">BigWig</span></a></footer></article></body></html>
