<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>BED · GenomicFeatures.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>GenomicFeatures.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><a class="toctext" href="../../intervals/">Intervals</a></li><li><span class="toctext">I/O</span><ul><li class="current"><a class="toctext" href>BED</a><ul class="internal"><li><a class="toctext" href="#Description-1">Description</a></li><li><a class="toctext" href="#Examples-1">Examples</a></li><li><a class="toctext" href="#API-1">API</a></li></ul></li><li><a class="toctext" href="../gff3/">GFF3</a></li><li><a class="toctext" href="../bigwig/">BigWig</a></li><li><a class="toctext" href="../bigbed/">BigBed</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>I/O</li><li><a href>BED</a></li></ul></nav><hr/><div id="topbar"><span>BED</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="BED-1" href="#BED-1">BED</a></h1><h2><a class="nav-anchor" id="Description-1" href="#Description-1">Description</a></h2><p>BED is a text-based file format for representing genomic annotations like genes, transcripts, and so on. A BED file has tab-delimited and variable-length fields; the first three fields denoting a genomic interval are mandatory.</p><p>This is an example of RNA transcripts:</p><pre><code class="language-none">chr9	68331023	68424451	NM_015110	0	+
chr9	68456943	68486659	NM_001206	0	-</code></pre><p>I/O tools for BED are provided from the <code>GenomicFeatures.BED</code> module, which exports following three types:</p><ul><li>Reader type: <code>BED.Reader</code></li><li>Writer type: <code>BED.Writer</code></li><li>Element type: <code>BED.Record</code></li></ul><h2><a class="nav-anchor" id="Examples-1" href="#Examples-1">Examples</a></h2><p>Here is a common workflow to iterate over all records in a BED file:</p><pre><code class="language-julia"># Import the BED module.
using GenomicFeatures

# Open a BED file.
reader = open(BED.Reader, &quot;data.bed&quot;)

# Iterate over records.
for record in reader
    # Do something on record (see Accessors section).
    chrom = BED.chrom(record)
    # ...
end

# Finally, close the reader.
close(reader)</code></pre><p>If you repeatedly access records within specific ranges, it would be more efficient to construct an <code>IntervalCollection</code> object from a BED reader:</p><pre><code class="language-julia"># Create an interval collection in memory.
icol = open(BED.Reader, &quot;data.bed&quot;) do reader
    IntervalCollection(reader)
end

# Query overlapping records.
for interval in eachoverlap(icol, Interval(&quot;chrX&quot;, 40001, 51500))
    # A record is stored in the metadata field of an interval.
    record = metadata(interval)
    # ...
end</code></pre><h2><a class="nav-anchor" id="API-1" href="#API-1">API</a></h2><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.Reader" href="#GenomicFeatures.BED.Reader"><code>GenomicFeatures.BED.Reader</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">BED.Reader(input::IO; index=nothing)
BED.Reader(input::AbstractString; index=:auto)</code></pre><p>Create a data reader of the BED file format.</p><p>The first argument specifies the data source. When it is a filepath that ends with <em>.bgz</em>, it is considered to be block compression file format (BGZF) and the function will try to find a tabix index file (&lt;filename&gt;.tbi) and read it if any. See <a href="http://www.htslib.org/doc/tabix.html">http://www.htslib.org/doc/tabix.html</a> for bgzip and tabix tools.</p><p><strong>Arguments</strong></p><ul><li><code>input</code>: data source</li><li><code>index</code>: path to a tabix file</li></ul></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.Writer" href="#GenomicFeatures.BED.Writer"><code>GenomicFeatures.BED.Writer</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">BED.Writer(output::IO)</code></pre><p>Create a data writer of the BED file format.</p><p><strong>Arguments:</strong></p><ul><li><code>output</code>: data sink</li></ul></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.Record" href="#GenomicFeatures.BED.Record"><code>GenomicFeatures.BED.Record</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">BED.Record()</code></pre><p>Create an unfilled BED record.</p></div></div><div><div><pre><code class="language-none">BED.Record(data::Vector{UInt8})</code></pre><p>Create a BED record object from <code>data</code>.</p><p>This function verifies and indexes fields for accessors. Note that the ownership of <code>data</code> is transferred to a new record object.</p></div></div><div><div><pre><code class="language-none">BED.Record(str::AbstractString)</code></pre><p>Create a BED record object from <code>str</code>.</p><p>This function verifies and indexes fields for accessors.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.chrom" href="#GenomicFeatures.BED.chrom"><code>GenomicFeatures.BED.chrom</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">chrom(record::Record)::String</code></pre><p>Get the chromosome name of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.chromstart" href="#GenomicFeatures.BED.chromstart"><code>GenomicFeatures.BED.chromstart</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">chromstart(record::Record)::Int</code></pre><p>Get the starting position of <code>record</code>.</p><p>Note that the first base is numbered 1.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.chromend" href="#GenomicFeatures.BED.chromend"><code>GenomicFeatures.BED.chromend</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">chromend(record::Record)::Int</code></pre><p>Get the end position of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.name" href="#GenomicFeatures.BED.name"><code>GenomicFeatures.BED.name</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">name(record::Record)::String</code></pre><p>Get the name of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.score" href="#GenomicFeatures.BED.score"><code>GenomicFeatures.BED.score</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">score(record::Record)::Int</code></pre><p>Get the score between 0 and 1000.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.strand" href="#GenomicFeatures.BED.strand"><code>GenomicFeatures.BED.strand</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">strand(record::Record)::GenomicFeatures.Strand</code></pre><p>Get the strand of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.thickstart" href="#GenomicFeatures.BED.thickstart"><code>GenomicFeatures.BED.thickstart</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">thickstart(record::Record)::Int</code></pre><p>Get the starting position at which <code>record</code> is drawn thickly.</p><p>Note that the first base is numbered 1.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.thickend" href="#GenomicFeatures.BED.thickend"><code>GenomicFeatures.BED.thickend</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">thickend(record::Record)::Int</code></pre><p>Get the end position at which <code>record</code> is drawn thickly.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.itemrgb" href="#GenomicFeatures.BED.itemrgb"><code>GenomicFeatures.BED.itemrgb</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">itemrgb(record::Record)::ColorTypes.RGB</code></pre><p>Get the RGB value of <code>record</code>.</p><p>The return type is defined in <a href="https://github.com/JuliaGraphics/ColorTypes.jl">ColorTypes.jl</a>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.blockcount" href="#GenomicFeatures.BED.blockcount"><code>GenomicFeatures.BED.blockcount</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">blockcount(record::Record)::Int</code></pre><p>Get the number of blocks (exons) in <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.blocksizes" href="#GenomicFeatures.BED.blocksizes"><code>GenomicFeatures.BED.blocksizes</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">blocksizes(record::Record)::Vector{Int}</code></pre><p>Get the block (exon) sizes of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicFeatures.BED.blockstarts" href="#GenomicFeatures.BED.blockstarts"><code>GenomicFeatures.BED.blockstarts</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">blockstarts(record::Record)::Vector{Int}</code></pre><p>Get the block (exon) starts of <code>record</code>.</p><p>Note that the first base is numbered 1.</p></div></div></section><footer><hr/><a class="previous" href="../../intervals/"><span class="direction">Previous</span><span class="title">Intervals</span></a><a class="next" href="../gff3/"><span class="direction">Next</span><span class="title">GFF3</span></a></footer></article></body></html>
