<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Sequence Composition · BioSequences.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../index.html"><img class="logo" src="../assets/logo.png" alt="BioSequences.jl logo"/></a><h1>BioSequences.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../">Home</a></li><li><a class="toctext" href="../symbols/">Biological Symbols</a></li><li><span class="toctext">Sequence types</span><ul><li><a class="toctext" href="../sequences/sequences/">Overview</a></li><li><a class="toctext" href="../sequences/bioseq/">BioSequence</a></li><li><a class="toctext" href="../sequences/refseq/">Reference Sequences</a></li><li><a class="toctext" href="../sequences/kmer/">Nucleic acid k-mers</a></li></ul></li><li><span class="toctext">IO</span><ul><li><a class="toctext" href="../io/fasta/">FASTA formatted files</a></li><li><a class="toctext" href="../io/fastq/">FASTQ formatted files</a></li><li><a class="toctext" href="../io/twobit/">2bit formatted files</a></li></ul></li><li><a class="toctext" href="../search/">Searching</a></li><li class="current"><a class="toctext" href>Sequence Composition</a><ul class="internal"></ul></li><li><a class="toctext" href="../demultiplexer/">Demultiplexing</a></li><li><a class="toctext" href="../contributing/">Contributing</a></li></ul></nav><article id="docs"><header><nav><ul><li><a href>Sequence Composition</a></li></ul></nav><hr/><div id="topbar"><span>Sequence Composition</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Sequence-composition-1" href="#Sequence-composition-1">Sequence composition</a></h1><p>There are many instances in analyzing sequence data where you will want to know about the composition of your sequences.</p><p>For example, for a given sequence, you may want to count how many of each possible Kmer, is present in the sequence. This would be important if - for instance - you wanted to analyze the Kmer spectra of your data. Alternatively you might have a collection of sequences, and may want to count how many of each unique sequence you have in your collection. This would be important if - for instance - your collection of sequences were from a population sample, and you wanted to compute the allele or genotype frequencies for the population.</p><p>Whatever the application, BioSequences provides a method called <code>composition</code>, and a parametric struct called <code>Composition</code> to both compute, and handle the results of such sequence composition calculations.</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.Composition" href="#BioSequences.Composition"><code>BioSequences.Composition</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Sequence composition.</p><p>This is a subtype of <code>Associative{T,Int}</code>, and the <code>getindex</code> method returns the number of occurrences of a symbol or a k-mer.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.composition" href="#BioSequences.composition"><code>BioSequences.composition</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">composition(seq | kmer_iter)</code></pre><p>Calculate composition of biological symbols in <code>seq</code> or k-mers in <code>kmer_iter</code>.</p></div></div><div><div><pre><code class="language-none">composition(iter)</code></pre><p>A generalised composition algorithm, which computes the number of unique items produced by an iterable.</p><p><strong>Example</strong></p><pre><code class="language-jlcon">
# Example, counting unique sequences.

julia&gt; a = dna&quot;AAAAAAAATTTTTT&quot;
14nt DNA Sequence:
AAAAAAAATTTTTT

julia&gt; b = dna&quot;AAAAAAAATTTTTT&quot;
14nt DNA Sequence:
AAAAAAAATTTTTT

julia&gt; c = a[5:10]
6nt DNA Sequence:
AAAATT

julia&gt; composition([a, b, c])
Vector{BioSequences.BioSequence{BioSequences.DNAAlphabet{4}}} Composition:
  AAAATT         =&gt; 1
  AAAAAAAATTTTTT =&gt; 2</code></pre></div></div></section><p>For example to get the nucleotide composition of a sequence:</p><pre><code class="language-julia-repl">julia&gt; comp = composition(dna&quot;ACGAG&quot;);

julia&gt; comp[DNA_A]
2

julia&gt; comp[DNA_T]
0
</code></pre><p>Composition structs behave like an associative collection, such as a <code>Dict</code>. But there are a few differences:</p><ol><li>The <code>getindex</code> method for Composition structs is overloaded to return a default value of 0, if a key is used that is not present in the Composition.</li><li>The <code>merge!</code> method for two Composition structs adds counts together, unlike the <code>merge!</code> method for other associative containers, which would overwrite the counts.</li></ol><p><code>merge!</code> is used to accumulate composition statistics of multiple sequences:</p><pre><code class="language-julia-repl">julia&gt; # initiaize an empty composition counter
       comp = composition(dna&quot;&quot;);
ERROR: LoadError: UndefVarError: @dna_str not defined
in expression starting at none:2

julia&gt; # iterate over sequences and accumulate composition statistics into `comp`
       for seq in seqs
           merge!(comp, composition(seq))
       end
ERROR: UndefVarError: seqs not defined

julia&gt; # or functional programming style in one line
       foldl((x, y) -&gt; merge(x, composition(y)), composition(dna&quot;&quot;), seqs)
ERROR: LoadError: UndefVarError: @dna_str not defined
in expression starting at none:3</code></pre><p><code>composition</code> is also applicable to a <em>k</em>-mer iterator:</p><pre><code class="language-julia-repl">julia&gt; comp = composition(each(DNAKmer{4}, dna&quot;ACGT&quot;^100));

julia&gt; comp[DNAKmer(&quot;ACGT&quot;)]
100

julia&gt; comp[DNAKmer(&quot;CGTA&quot;)]
99
</code></pre><footer><hr/><a class="previous" href="../search/"><span class="direction">Previous</span><span class="title">Searching</span></a><a class="next" href="../demultiplexer/"><span class="direction">Next</span><span class="title">Demultiplexing</span></a></footer></article></body></html>
