<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Genetic Diversity · GeneticVariation.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="GeneticVariation.jl logo"/></a><h1>GeneticVariation.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><span class="toctext">Manual</span><ul><li><span class="toctext">I/O</span><ul><li><a class="toctext" href="../io/vcf-bcf/">VCF and BCF formatted files</a></li></ul></li><li><a class="toctext" href="../site_counting/">Mutation Counting</a></li><li class="current"><a class="toctext" href>Genetic Diversity</a><ul class="internal"><li class="toplevel"><a class="toctext" href="#Computing-measures-of-genetic-diversity-1">Computing measures of genetic diversity</a></li><li><a class="toctext" href="#Nucleotide-diversity-1">Nucleotide diversity</a></li></ul></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>Manual</li><li><a href>Genetic Diversity</a></li></ul></nav><hr/><div id="topbar"><span>Genetic Diversity</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Computing-allele-frequencies-1" href="#Computing-allele-frequencies-1">Computing allele frequencies</a></h1><p>When first looking at the diversity present in a population, it is common to want to know how many of each unique allele there is in a population i.e. the allele frequencies of the population are.</p><p>Formally defined, allele frequency is a measure of the relative frequency of an allele on a genetic locus in a population.</p><p>In population genetics, allele frequencies show the genetic diversity of a species population or equivalently the richness of its gene pool.</p><p>Population genetics studies the different &quot;forces&quot; that might lead to changes in the distribution and frequencies of alleles - in other words, to evolution.</p><p>Besides selection, these forces include genetic drift, mutation and migration.</p><p>Computing allele frequencies then, is an essential task for many wishing to work with genetic variation, and so methods for computing such frequencies are included in GeneticVariation.jl.</p><p>Allele frequencies can be computed for genes, micro-satellites, RFPL patterns, and from SNPs.</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeneticVariation.gene_frequencies" href="#GeneticVariation.gene_frequencies"><code>GeneticVariation.gene_frequencies</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">gene_frequencies(seqcounts::Composition{T}) where T &lt;: Sequence</code></pre><p>Compute gene_frequencies from a <code>BioSequences.Composition</code> variable that contains unique sequence counts.</p></div></div><div><div><pre><code class="language-none">gene_frequencies(iterable)</code></pre><p>Compute the gene frequencies for any iterable with an <code>eltype</code> which is a concrete subtype of the abstract <code>Sequence</code> type.</p></div></div></section><h1><a class="nav-anchor" id="Computing-measures-of-genetic-diversity-1" href="#Computing-measures-of-genetic-diversity-1">Computing measures of genetic diversity</a></h1><p>There are various methods of quantifying the amount of genetic variation in biological data with GeneticVariation.jl:</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeneticVariation.avg_mut" href="#GeneticVariation.avg_mut"><code>GeneticVariation.avg_mut</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">avg_mut(sequences)</code></pre><p>The average number of mutations found in (n choose 2) pairwise comparisons of sequences (i, j) in a sample of sequences.</p><p><code>sequences</code> should be any indexable container of DNA sequence types.</p></div></div></section><h2><a class="nav-anchor" id="Nucleotide-diversity-1" href="#Nucleotide-diversity-1">Nucleotide diversity</a></h2><p>Nucleotide diversity is a concept in molecular genetics which is used to measure the degree of polymorphism within a population.</p><p>There are different methods which can be used to compute measures of nucleotide diversity, we list them below, and show how to compute them using GeneticVariation.</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GeneticVariation.NL79" href="#GeneticVariation.NL79"><code>GeneticVariation.NL79</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">NL79(m::M, f::V) where {M&lt;:AbstractMatrix{Float64},V&lt;:AbstractVector{Float64}}</code></pre><p>Compute nucleotide diversity using a matrix of the number of mutations between sequence pairs, and a vector of the frequencies of each sequence in the population.</p></div></div><div><div><pre><code class="language-none">NL79(sequences)</code></pre><p>Compute nucleotide diversity, as described by Nei and Li in 1979.</p><p>This measure is defined as the average number of nucleotide differences per site between two DNA sequences in all possible pairs in the sample population, and is often denoted by the greek letter pi.</p><p><code>Sequences</code> should be any iterable that yields biosequence types.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; testSeqs = [dna&quot;AAAACTTTTACCCCCGGGGG&quot;,
                   dna&quot;AAAACTTTTACCCCCGGGGG&quot;,
                   dna&quot;AAAACTTTTACCCCCGGGGG&quot;,
                   dna&quot;AAAACTTTTACCCCCGGGGG&quot;,
                   dna&quot;AAAAATTTTACCCCCGTGGG&quot;,
                   dna&quot;AAAAATTTTACCCCCGTGGG&quot;,
                   dna&quot;AAAACTTTTTCCCCCGTAGG&quot;,
                   dna&quot;AAAACTTTTTCCCCCGTAGG&quot;,
                   dna&quot;AAAAATTTTTCCCCCGGAGG&quot;,
                   dna&quot;AAAAATTTTTCCCCCGGAGG&quot;]
10-element Array{BioSequences.BioSequence{BioSequences.DNAAlphabet{4}},1}:
 AAAACTTTTACCCCCGGGGG
 AAAACTTTTACCCCCGGGGG
 AAAACTTTTACCCCCGGGGG
 AAAACTTTTACCCCCGGGGG
 AAAAATTTTACCCCCGTGGG
 AAAAATTTTACCCCCGTGGG
 AAAACTTTTTCCCCCGTAGG
 AAAACTTTTTCCCCCGTAGG
 AAAAATTTTTCCCCCGGAGG
 AAAAATTTTTCCCCCGGAGG

 julia&gt; NL79(testSeqs)
 0.096
</code></pre></div></div></section><footer><hr/><a class="previous" href="../site_counting/"><span class="direction">Previous</span><span class="title">Mutation Counting</span></a></footer></article></body></html>
