<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Workflow Â· JWAS.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="JWAS.jl logo"/></a><h1>JWAS.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><a class="toctext" href="../../theory/theory/">Some Theory</a></li><li><a class="toctext" href="../../citing/citing/">Citing</a></li><li><span class="toctext">Manual</span><ul><li><a class="toctext" href="../getstarted/">Get Started</a></li><li class="current"><a class="toctext" href>Workflow</a><ul class="internal"><li><a class="toctext" href="#Available-Models-1">Available Models</a></li><li><a class="toctext" href="#Get-Data-Ready-1">Get Data Ready</a></li><li><a class="toctext" href="#Step-1:-Load-Packages-1">Step 1: Load Packages</a></li><li><a class="toctext" href="#Step-2:-Read-data-1">Step 2: Read data</a></li><li><a class="toctext" href="#Step-3:-Build-Model-Equations-1">Step 3: Build Model Equations</a></li><li><a class="toctext" href="#Step-4:-Set-Factors-or-Covariate-1">Step 4: Set Factors or Covariate</a></li><li><a class="toctext" href="#Step-5:-Set-Random-or-Fixed-Effects-1">Step 5: Set Random or Fixed Effects</a></li><li><a class="toctext" href="#Step-6:-Use-Genomic-Information-1">Step 6: Use Genomic Information</a></li><li><a class="toctext" href="#Step-7:-Run-Bayesian-Analysis-1">Step 7: Run Bayesian Analysis</a></li><li><a class="toctext" href="#check-results-1">check results</a></li></ul></li><li><a class="toctext" href="../public/">Public</a></li><li><a class="toctext" href="../internals/">Internals</a></li></ul></li><li><span class="toctext">Examples</span><ul><li><a class="toctext" href="../../examples/examples/">Examples</a></li></ul></li><li><a class="toctext" href="../../FrequentlyAskedQuestions/FrequentlyAskedQuestions/">Frequently Asked Questions</a></li></ul></nav><article id="docs"><header><nav><ul><li>Manual</li><li><a href>Workflow</a></li></ul></nav><hr/><div id="topbar"><span>Workflow</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Workflow-1" href="#Workflow-1">Workflow</a></h1><p>A step by step workflow for how to run JWAS is shown in this section. The workflow below is used to demonstrate a three-trait Bayesian linear mixed model fitting fixed effects (x1, x3), random effects (x2), direct genetic effects (ID), maternal genetic effects (dam) and genomic information.</p><ul><li><a href="#Workflow-1">Workflow</a></li><ul><li><a href="#Available-Models-1">Available Models</a></li><li><a href="#Get-Data-Ready-1">Get Data Ready</a></li><li><a href="#Step-1:-Load-Packages-1">Step 1: Load Packages</a></li><li><a href="#Step-2:-Read-data-1">Step 2: Read data</a></li><li><a href="#Step-3:-Build-Model-Equations-1">Step 3: Build Model Equations</a></li><li><a href="#Step-4:-Set-Factors-or-Covariate-1">Step 4: Set Factors or Covariate</a></li><li><a href="#Step-5:-Set-Random-or-Fixed-Effects-1">Step 5: Set Random or Fixed Effects</a></li><li><a href="#Step-6:-Use-Genomic-Information-1">Step 6: Use Genomic Information</a></li><li><a href="#Step-7:-Run-Bayesian-Analysis-1">Step 7: Run Bayesian Analysis</a></li><li><a href="#check-results-1">check results</a></li></ul></ul><h2><a class="nav-anchor" id="Available-Models-1" href="#Available-Models-1">Available Models</a></h2><p>Given the data and model equations, several different types of models are available in JWAS as shown below. In the table below, &quot;X&quot; denotes the type of available data, and &quot;Y&lt;=A&quot; denotes that Y individuals is a subset of A individuals.  </p><table><tr><th>Linear Mixed Models (LMM)</th><th>phenotypes (Y)</th><th>pedigree (A)</th><th>genotypes (G)</th><th>notes</th></tr><tr><td>Conventional LMM</td><td>X</td><td></td><td></td><td></td></tr><tr><td>Pedigree-based LMM</td><td>X</td><td>X</td><td></td><td>Y&lt;=A</td></tr><tr><td>Complete Genomic LMM</td><td>X</td><td>maybe</td><td>X</td><td>Y&lt;=G</td></tr><tr><td>Incomplete Genomic LMM</td><td>X</td><td>X</td><td>X</td><td>Y&lt;=A,G&lt;=A</td></tr></table><div class="admonition note"><div class="admonition-title">Note</div><div class="admonition-text"><ul><li><p><strong>Incomplete Genomic LMM</strong> is also called &quot;single-step&quot; methods in animal breeding.</p></li><li><p>Pedigree information may be used in <strong>Complete Genomic LMM</strong> for extra polygenic effects to account for genetic variance not explained by the genomic data (e.g., SNPs).</p></li><li><p><strong>Pedigree-based LMM</strong> (none of the individuals in the pedigree are genotyped) and <strong>Complete Genomic LMM</strong> (all individuals in the pedigree are genotyped) are special cases of <strong>Incomplete Genomic LMM</strong> (part of the individuals in the pedigree are genotyped).</p></li></ul></div></div><h2><a class="nav-anchor" id="Get-Data-Ready-1" href="#Get-Data-Ready-1">Get Data Ready</a></h2><p>By default, input data files are comma-separated values (CSV) files, where each line of the file consists of one or more fields, separated by <strong>commas</strong>. Other field separators such as space (&#39; &#39;) or tab (&#39;\t&#39;) can be used if you supply the keyword argument, e.g, <code>CSV.read(...,delim=&#39;\t&#39;)</code> or <code>add_genotypes(...,separator=&#39;\t&#39;)</code></p><p>Click on the buttons inside the tabbed menu to see the data:</p><head>
<meta name="viewport" content="width=device-width, initial-scale=1">
<style>
body {font-family: Arial;}

/* Style the tab */
.tab {
    overflow: hidden;
    border: 1px solid #ccc;
    background-color: #f1f1f1;
}

/* Style the buttons inside the tab */
.tab button {
    background-color: inherit;
    float: left;
    border: none;
    outline: none;
    cursor: pointer;
    padding: 14px 16px;
    transition: 0.3s;
    font-size: 17px;
}

/* Change background color of buttons on hover */
.tab button:hover {
    background-color: #ddd;
}

/* Create an active/current tablink class */
.tab button.active {
    background-color: #ccc;
}

/* Style the tab content */
.tabcontent {
    display: none;
    padding: 6px 12px;
    border: 1px solid #ccc;
    border-top: none;
}
</style>
</head>
<body>

<div class="tab">
  <button class="tablinks" onclick="openCity(event, 'phenotypes')">phenotypes.txt</button>
  <button class="tablinks" onclick="openCity(event, 'pedigree')">pedigree.txt</button>
  <button class="tablinks" onclick="openCity(event, 'genotypes')">genotypes.txt</button>
  <button class="tablinks" onclick="openCity(event, 'map file')">map.txt</button>
</div>

<div id="phenotypes" class="tabcontent">
<p>ID,y1,y2,y3,x1,x2,x3,dam</p>
<p>a1,-0.06,3.58,-1.18,0.9,2,m,0</p>
<p>a2,-0.6,4.9,0.88,0.3,1,f,0</p>
<p>a3,-2.07,3.19,0.73,0.7,2,f,0</p>
<p>a4,-2.63,6.97,-0.83,0.6,1,m,a2</p>
<p>a5,2.31,3.5,-1.52,0.4,2,m,a2</p>
<p>a6,0.93,4.87,-0.01,05,2,f,a3</p>
<p>a7,-0.69,3.1,-1.47,0.5,2,f,a3</p>
<p>a8,-4.69,7.31,-1.09,0.3,2,m,a6</p>
<p>a9,-2.81,7.18,0.76,0.4,2,m,a6</p>
<p>a10,1.92,1.78,-0.88,0.2,1,m,a7</p>
</div>

<div id="pedigree" class="tabcontent">
<p>ID,Sire,Dam</p>
<p>a1,0,0</p>
<p>a2,0,0</p>
<p>a3,0,0</p>
<p>a4,a1,a2</p>
<p>a5,a1,a2</p>
<p>a6,a1,a3</p>
<p>a7,a1,a3</p>
<p>a8,a4,a6</p>
<p>a9,a4,a6</p>
<p>a10,a5,a7</p>
</div>

<div id="genotypes" class="tabcontent">
<p>ID,m1,m2,m3,m4,m5</p>
<p>a1,1,2,1,1,0</p>
<p>a2,2,1,1,1,1</p>
<p>a3,1,1,0,1,1</p>
<p>a4,2,2,0,1,0</p>
<p>a5,1,1,2,1,1</p>
<p>a6,2,1,0,0,0</p>
<p>a7,0,2,1,2,1</p>
<p>a8,2,2,0,0,0</p>
<p>a9,2,1,0,1,0</p>
<p>a10,0,2,2,2,1</p>
</div>

<div id="map file" class="tabcontent">
<p>markerID,chromosome,position</p>
<p>m1,1,16977</p>
<p>m2,1,434311</p>
<p>m3,1,1025513</p>
<p>m4,2,70350</p>
<p>m5,2,101135</p>
</div>


<script>
function openCity(evt, cityName) {
    var i, tabcontent, tablinks;
    tabcontent = document.getElementsByClassName("tabcontent");
    for (i = 0; i < tabcontent.length; i++) {
        tabcontent[i].style.display = "none";
    }
    tablinks = document.getElementsByClassName("tablinks");
    for (i = 0; i < tablinks.length; i++) {
        tablinks[i].className = tablinks[i].className.replace(" active", "");
    }
    document.getElementById(cityName).style.display = "block";
    evt.currentTarget.className += " active";
}
</script>
</body><h2><a class="nav-anchor" id="Step-1:-Load-Packages-1" href="#Step-1:-Load-Packages-1">Step 1: Load Packages</a></h2><pre><code class="language-julia">using JWAS,CSV,DataFrames</code></pre><hr/><p>The <code>JWAS</code> package is loaded, as well as the <code>CSV</code> and <code>DataFrame</code> packages for reading text files.</p><h2><a class="nav-anchor" id="Step-2:-Read-data-1" href="#Step-2:-Read-data-1">Step 2: Read data</a></h2><pre><code class="language-julia">phenotypes = CSV.read(&quot;phenotypes.txt&quot;,delim = &#39;,&#39;,header=true)
pedigree   = get_pedigree(&quot;pedigree.txt&quot;,separator=&quot;,&quot;,header=true)
head(phenotypes)</code></pre><p>output:</p><pre><code class="language-julia">6Ã8 DataFrames.DataFrame
â Row â ID â y1    â y2   â y3    â x1  â x2 â x3 â dam â
âââââââ¼âââââ¼ââââââââ¼âââââââ¼ââââââââ¼ââââââ¼âââââ¼âââââ¼ââââââ¤
â 1   â a1 â -0.06 â 3.58 â -1.18 â 0.9 â 2  â m  â 0   â
â 2   â a2 â -0.6  â 4.9  â 0.88  â 0.3 â 1  â f  â 0   â
â 3   â a3 â -2.07 â 3.19 â 0.73  â 0.7 â 2  â f  â 0   â
â 4   â a4 â -2.63 â 6.97 â -0.83 â 0.6 â 1  â m  â a2  â
â 5   â a5 â 2.31  â 3.5  â -1.52 â 0.4 â 2  â m  â a2  â
â 6   â a6 â 0.93  â 4.87 â -0.01 â 5.0 â 2  â f  â a3  â</code></pre><ul><li>link to documentation for <a href="../public/#JWAS.PedModule.get_pedigree"><code>get_pedigree</code></a></li></ul><hr/><p>The <strong>phenotypic data</strong> is read on line 1, and the <strong>pedigree data</strong> is read on line 2. On line 3, the first several rows of data are shown.</p><h2><a class="nav-anchor" id="Step-3:-Build-Model-Equations-1" href="#Step-3:-Build-Model-Equations-1">Step 3: Build Model Equations</a></h2><pre><code class="language-julia">model_equation = &quot;y1 = intercept + x1 + x3 + ID + dam
                  y2 = intercept + x1 + x2 + x3 + ID  
                  y3 = intercept + x1 + x1*x3 + x2 + ID&quot;
model=build_model(model_equation, R)</code></pre><ul><li>link to documentation for <a href="../public/#JWAS.build_model"><code>build_model</code></a></li></ul><hr/><p>The non-genomic part of the model equation for a 3-trait analysis is defined on the first 3 lines.</p><ul><li>The effects fitted in the model for trait <code>y1</code> are the intercept, <code>x1</code>, <code>x3</code>, direct genetic effects (<code>ID</code>) and maternal genetic effects (<code>dam</code>).</li><li>The effects fitted in the model for trait <code>y2</code> are the intercept, <code>x1</code>, <code>x2</code>, <code>x3</code> and direct genetic effects (<code>ID</code>).</li><li>The effects fitted in the model for trait <code>y3</code> are the intercept, <code>x1</code>, the interaction between <code>x1</code> and <code>x3</code>, <code>x2</code> and direct genetic effects (<code>ID</code>).</li></ul><p>On the last line, the model is built given the model equation and residual variance <code>R</code> (a 3x3 matrix). By default, all effects are treated as fixed and classed as factors (categorical variables) rather than covariates (quantitative variables).</p><h2><a class="nav-anchor" id="Step-4:-Set-Factors-or-Covariate-1" href="#Step-4:-Set-Factors-or-Covariate-1">Step 4: Set Factors or Covariate</a></h2><pre><code class="language-julia">set_covariate(model,&quot;x1&quot;)</code></pre><ul><li>link to documentation for <a href="../public/#JWAS.set_covariate"><code>set_covariate</code></a></li></ul><hr/><p>On line 1, the effect <code>x1</code> is defined to be a covariate rather than class effect.</p><h2><a class="nav-anchor" id="Step-5:-Set-Random-or-Fixed-Effects-1" href="#Step-5:-Set-Random-or-Fixed-Effects-1">Step 5: Set Random or Fixed Effects</a></h2><pre><code class="language-julia">set_random(model,&quot;x2&quot;,G1)
set_random(model,&quot;ID dam&quot;,pedigree,G2)</code></pre><ul><li>link to documentation for <a href="../public/#JWAS.set_random"><code>set_random</code></a></li></ul><hr/><p>On line 1, the <code>x2</code> class effect is defined as random with variance <code>G1</code>(a 2x2 matrix). On line 2, direct genetic effects and maternal genetic effects are fitted as <code>ID</code> and <code>dam</code> with <code>G2</code> (a 4x4 matrix) and the inverse of the numerator relationship matrix defined from pedigree.</p><h2><a class="nav-anchor" id="Step-6:-Use-Genomic-Information-1" href="#Step-6:-Use-Genomic-Information-1">Step 6: Use Genomic Information</a></h2><pre><code class="language-julia">add_genotypes(model,&quot;genotypes.txt&quot;,G3,separator=&#39;,&#39;)</code></pre><ul><li>link to documentation for <a href="../public/#JWAS.add_genotypes"><code>add_genotypes</code></a></li></ul><hr/><p>On line 1, the genomic part of the model is defined with the genotype file and variance <code>G3</code> (a 3x3 matrix).</p><h2><a class="nav-anchor" id="Step-7:-Run-Bayesian-Analysis-1" href="#Step-7:-Run-Bayesian-Analysis-1">Step 7: Run Bayesian Analysis</a></h2><pre><code class="language-julia">outputMCMCsamples(model,&quot;x2&quot;)
out=runMCMC(model,phenotypes,methods=&quot;BayesC&quot;,output_samples_frequency=100)</code></pre><ul><li>link to documentation for <a href="../public/#JWAS.outputMCMCsamples"><code>outputMCMCsamples</code></a></li><li>link to documentation for <a href="../public/#JWAS.runMCMC"><code>runMCMC</code></a></li></ul><hr/><p>On line 1, MCMC samples from <code>runMCMC</code> for <code>x2</code> is saved to a file, where each row represents one sample from the MCMC. On line 2, a multi-trait BayesC analysis is performed with <code>model</code> and <code>phenotypes</code> as had been defined in step 1-6. MCMC samples for marker effects, location parameters specified on line 1, and all variance components from this analysis are saved every <code>output_samples_frequency</code> iterations to files.</p><hr/><p>Several steps above can be skipped if no related information is available, e.g., step 6 is skipped for pedigree-based LMM. Several detailed examples are available in the examples section. Here is the link to documentation for all <a href="../public/#Public-functions-1">Public functions</a>.</p><h2><a class="nav-anchor" id="check-results-1" href="#check-results-1">check results</a></h2><p>Posterior means of location parameters, most variance components, and marker effects are saved in <code>out</code>. They can be listed and obtained as</p><pre><code class="language-julia">keys(out)

# output:
#
# Base.KeyIterator for a Dict{Any,Any} with 7 entries. Keys:
#   &quot;Posterior mean of polygenic effects covariance matrix&quot;
#   &quot;Model frequency&quot;
#   &quot;Posterior mean of residual covariance matrix&quot;
#   &quot;Posterior mean of marker effects&quot;
#   &quot;Posterior mean of marker effects covariance matrix&quot;
#   &quot;Posterior mean of location parameters&quot;
#   &quot;Posterior mean of Pi&quot;

out[&quot;Posterior mean of residual covariance matrix&quot;]

# output:
#
# 3Ã3 Array{Float64,2}:
#   0.674651   -0.103877   0.0834044
#  -0.103877    0.828135  -0.121798
#   0.0834044  -0.121798   0.720751
</code></pre><p>MCMC samples for marker effects, location parameters specified in step 7, and all variance components are saved to text files in your working directory. They can be obtained as</p><pre><code class="language-julia">res=readdlm(&quot;MCMC_samples_marker_effects_y1.txt&quot;,&#39;,&#39;,header=true)</code></pre><footer><hr/><a class="previous" href="../getstarted/"><span class="direction">Previous</span><span class="title">Get Started</span></a><a class="next" href="../public/"><span class="direction">Next</span><span class="title">Public</span></a></footer></article></body></html>
