<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>BioSequence · BioSequences.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="BioSequences.jl logo"/></a><h1>BioSequences.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><a class="toctext" href="../../symbols/">Biological Symbols</a></li><li><span class="toctext">Sequence types</span><ul><li><a class="toctext" href="../sequences/">Overview</a></li><li class="current"><a class="toctext" href>BioSequence</a><ul class="internal"><li><a class="toctext" href="#Constructing-sequences-1">Constructing sequences</a></li><li><a class="toctext" href="#Indexing,-modifying-and-transformations-1">Indexing, modifying and transformations</a></li><li><a class="toctext" href="#Site-counting-1">Site counting</a></li><li><a class="toctext" href="#Iteration-1">Iteration</a></li><li><a class="toctext" href="#Using-a-more-compact-sequence-representation-1">Using a more compact sequence representation</a></li><li><a class="toctext" href="#Defining-a-new-alphabet-1">Defining a new alphabet</a></li></ul></li><li><a class="toctext" href="../refseq/">Reference Sequences</a></li><li><a class="toctext" href="../kmer/">Nucleic acid k-mers</a></li></ul></li><li><span class="toctext">IO</span><ul><li><a class="toctext" href="../../io/fasta/">FASTA formatted files</a></li><li><a class="toctext" href="../../io/fastq/">FASTQ formatted files</a></li><li><a class="toctext" href="../../io/twobit/">2bit formatted files</a></li></ul></li><li><a class="toctext" href="../../search/">Searching</a></li><li><a class="toctext" href="../../composition/">Sequence Composition</a></li><li><a class="toctext" href="../../demultiplexer/">Demultiplexing</a></li><li><a class="toctext" href="../../contributing/">Contributing</a></li></ul></nav><article id="docs"><header><nav><ul><li>Sequence types</li><li><a href>BioSequence</a></li></ul></nav><hr/><div id="topbar"><span>BioSequence</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="General-purpose-sequences-1" href="#General-purpose-sequences-1">General-purpose sequences</a></h1><p><code>BioSequence{A}</code> is a generic sequence type parameterized by an alphabet type <code>A</code> that defines the domain (or set) of biological symbols, and each alphabet has an associated symbol type. For example, <code>AminoAcidAlphabet</code> is associated with <code>AminoAcid</code> and hence an object of the <code>BioSequence{AminoAcidAlphabet}</code> type represents a sequence of amino acids.  Symbols from multiple alphabets can&#39;t be intermixed in one sequence type.</p><p>The following table summarizes common sequence types that are defined in the <code>BioSequences</code> module:</p><table><tr><th>Type</th><th>Symbol type</th><th>Type alias</th></tr><tr><td><code>BioSequence{DNAAlphabet{4}}</code></td><td><code>DNA</code></td><td><code>DNASequence</code></td></tr><tr><td><code>BioSequence{RNAAlphabet{4}}</code></td><td><code>RNA</code></td><td><code>RNASequence</code></td></tr><tr><td><code>BioSequence{AminoAcidAlphabet}</code></td><td><code>AminoAcid</code></td><td><code>AminoAcidSequence</code></td></tr><tr><td><code>BioSequence{CharAlphabet}</code></td><td><code>Char</code></td><td><code>CharSequence</code></td></tr></table><p>Parameterized definition of the <code>BioSequence{A}</code> type is for the purpose of unifying the data structure and operations of any symbol type. In most cases, users don&#39;t have to care about it and can use <em>type aliases</em> listed above. However, the alphabet type fixes the internal memory encoding and plays an important role when optimizing performance of a program (see <a href="#Using-a-more-compact-sequence-representation-1">Using a more compact sequence representation</a> section for low-memory encodings).  It also enables a user to define their own alphabet only by defining few numbers of methods. This is described in <a href="#Defining-a-new-alphabet-1">Defining a new alphabet</a> section.</p><h2><a class="nav-anchor" id="Constructing-sequences-1" href="#Constructing-sequences-1">Constructing sequences</a></h2><h3><a class="nav-anchor" id="Using-string-literals-1" href="#Using-string-literals-1">Using string literals</a></h3><p>Most immediately, sequence literals can be constructed using the string macros <code>dna</code>, <code>rna</code>, <code>aa</code>, and <code>char</code>:</p><pre><code class="language-julia-repl">julia&gt; dna&quot;TACGTANNATC&quot;
11nt DNA Sequence:
TACGTANNATC

julia&gt; rna&quot;AUUUGNCCANU&quot;
11nt RNA Sequence:
AUUUGNCCANU

julia&gt; aa&quot;ARNDCQEGHILKMFPSTWYVX&quot;
21aa Amino Acid Sequence:
ARNDCQEGHILKMFPSTWYVX

julia&gt; char&quot;αβγδϵ&quot;
5char Char Sequence:
αβγδϵ
</code></pre><p>However it should be noted that by default these sequence literals allocate the <code>BioSequence</code> object before the code containing the sequence literal is run. This means there may be occasions where your program does not behave as you first expect. For example consider the following code:</p><pre><code class="language-julia-repl">julia&gt; function foo()
           s = dna&quot;CTT&quot;
           push!(s, DNA_A)
       end
foo (generic function with 1 method)
</code></pre><p>You might expect that every time you call <code>foo</code>, that a DNA sequence <code>CTTA</code> would be returned. You might expect that this is because every time <code>foo</code> is called, a new DNA sequence variable <code>CTT</code> is created, and and <code>A</code> nucleotide is pushed to it, and the result, <code>CTTA</code> is returned. In other words you might expect the following output:</p><pre><code class="language-julia-repl">julia&gt; foo()
4nt DNA Sequence:
CTTA

julia&gt; foo()
4nt DNA Sequence:
CTTA

julia&gt; foo()
4nt DNA Sequence:
CTTA
</code></pre><p>However, this is not what happens, instead the following happens:</p><pre><code class="language-julia-repl">julia&gt; foo()
4nt DNA Sequence:
CTTA

julia&gt; foo()
5nt DNA Sequence:
CTTAA

julia&gt; foo()
6nt DNA Sequence:
CTTAAA
</code></pre><p>The reason for this is because the sequence literal is allocated only once before the first time the function <code>foo</code> is called and run. Therefore, <code>s</code> in <code>foo</code> is always a reference to that one sequence that was allocated. So one sequence is created before <code>foo</code> is called, and then it is pushed to every time <code>foo</code> is called. Thus, that one allocated sequence grows with every call of <code>foo</code>.</p><p>If you wanted <code>foo</code> to create a new sequence each time it is called, then you can add a flag to the end of the sequence literal to dictate behaviour: A flag of &#39;s&#39; means &#39;static&#39;: the sequence will be allocated before code is run, as is the default behaviour described above. However providing &#39;d&#39; flag changes the behaviour: &#39;d&#39; means &#39;dynamic&#39;: the sequence will be allocated at whilst the code is running, and not before. So to change <code>foo</code> so as it creates a new sequence each time it is called, simply add the &#39;d&#39; flag to the sequence literal:</p><pre><code class="language-julia-repl">julia&gt; function foo()
           s = dna&quot;CTT&quot;d     # &#39;d&#39; flag appended to the string literal.
           push!(s, DNA_A)
       end
foo (generic function with 1 method)
</code></pre><p>Now every time <code>foo</code> is called, a new sequence <code>CTT</code> is created, and an <code>A</code> nucleotide is pushed to it:</p><pre><code class="language-julia-repl">julia&gt; foo()
4nt DNA Sequence:
CTTA

julia&gt; foo()
4nt DNA Sequence:
CTTA

julia&gt; foo()
4nt DNA Sequence:
CTTA
</code></pre><p>So the take home message of sequence literals is this:</p><p>Be careful when you are using sequence literals inside of functions, and inside the bodies of things like for loops. And if you use them and are unsure, use the  &#39;s&#39; and &#39;d&#39; flags to ensure the behaviour you get is the behaviour you intend.</p><h3><a class="nav-anchor" id="Other-constructors-and-conversion-1" href="#Other-constructors-and-conversion-1">Other constructors and conversion</a></h3><p>Sequences can also be constructed from strings or arrays of nucleotide or amino acid symbols using constructors or the <code>convert</code> function:</p><pre><code class="language-julia-repl">julia&gt; DNASequence(&quot;TTANC&quot;)
5nt DNA Sequence:
TTANC

julia&gt; DNASequence([DNA_T, DNA_T, DNA_A, DNA_N, DNA_C])
5nt DNA Sequence:
TTANC

julia&gt; convert(DNASequence, [DNA_T, DNA_T, DNA_A, DNA_N, DNA_C])
5nt DNA Sequence:
TTANC
</code></pre><p>Using <code>convert</code>, these operations are reversible: sequences can be converted to strings or arrays:</p><pre><code class="language-julia-repl">julia&gt; convert(String, dna&quot;TTANGTA&quot;)
&quot;TTANGTA&quot;

julia&gt; convert(Vector{DNA}, dna&quot;TTANGTA&quot;)
7-element Array{DNA,1}:
 DNA_T
 DNA_T
 DNA_A
 DNA_N
 DNA_G
 DNA_T
 DNA_A
</code></pre><p>Sequences can also be concatenated into longer sequences:</p><pre><code class="language-julia-repl">julia&gt; DNASequence(dna&quot;ACGT&quot;, dna&quot;NNNN&quot;, dna&quot;TGCA&quot;)
12nt DNA Sequence:
ACGTNNNNTGCA

julia&gt; dna&quot;ACGT&quot; * dna&quot;TGCA&quot;
8nt DNA Sequence:
ACGTTGCA

julia&gt; repeat(dna&quot;TA&quot;, 10)
20nt DNA Sequence:
TATATATATATATATATATA

julia&gt; dna&quot;TA&quot; ^ 10
20nt DNA Sequence:
TATATATATATATATATATA
</code></pre><p>Despite being separate types, <code>DNASequence</code> and <code>RNASequence</code> can freely be converted between efficiently without copying the underlying data:</p><pre><code class="language-julia-repl">julia&gt; dna = dna&quot;TTANGTAGACCG&quot;
12nt DNA Sequence:
TTANGTAGACCG

julia&gt; rna = convert(RNASequence, dna)
12nt RNA Sequence:
UUANGUAGACCG

julia&gt; dna.data === rna.data  # underlying data are same
true
</code></pre><p>A random sequence can be obtained by the <code>randdnaseq</code>, <code>randrnaseq</code> and <code>randaaseq</code> functions, which generate <code>DNASequence</code>, <code>RNASequence</code> and <code>AminoAcidSequence</code>, respectively. Generated sequences are composed of the standard symbols without ambiguity and gap. For example, <code>randdnaseq(6)</code> may generate <code>dna&quot;TCATAG&quot;</code> but never generates <code>dna&quot;TNANAG&quot;</code> or <code>dna&quot;T-ATAG&quot;</code>.</p><p>A translatable <code>RNASequence</code> can also be converted to an <code>AminoAcidSequence</code> using the <a href="#BioSequences.translate"><code>translate</code></a> function.</p><h2><a class="nav-anchor" id="Indexing,-modifying-and-transformations-1" href="#Indexing,-modifying-and-transformations-1">Indexing, modifying and transformations</a></h2><h3><a class="nav-anchor" id="Getindex-1" href="#Getindex-1">Getindex</a></h3><p>Sequences for the most part behave like other vector or string types. They can be indexed using integers or ranges:</p><pre><code class="language-julia-repl">julia&gt; seq = dna&quot;ACGTTTANAGTNNAGTACC&quot;
19nt DNA Sequence:
ACGTTTANAGTNNAGTACC

julia&gt; seq[5]
DNA_T

julia&gt; seq[6:end]
14nt DNA Sequence:
TANAGTNNAGTACC
</code></pre><p>Note that, indexing a biological sequence by range creates a subsequence of the original sequence. Unlike <code>Arrays</code> in the standard library, creating a subsequence is copy-free: a subsequence simply points to the original sequence data with its range. You may think that this is unsafe because modifying subsequences propagates to the original sequence, but this doesn&#39;t happen actually:</p><pre><code class="language-julia-repl">julia&gt; seq = dna&quot;AAAA&quot;    # create a sequence
4nt DNA Sequence:
AAAA

julia&gt; subseq = seq[1:2]  # create a subsequence from `seq`
2nt DNA Sequence:
AA

julia&gt; subseq[2] = DNA_T  # modify the second element of it
DNA_T

julia&gt; subseq             # the subsequence is modified
2nt DNA Sequence:
AT

julia&gt; seq                # but the original sequence is not
4nt DNA Sequence:
AAAA
</code></pre><p>This is because modifying a sequence checks whether its underlying data are shared with other sequences under the hood. If and only if the data are shared, the subsequence creates a copy of itself. Any modifying operation does this check. This is called <em>copy-on-write</em> strategy and users don&#39;t need to care about it because it is transparent: If the user modifies a sequence with or subsequence, the job of managing and protecting the underlying data of sequences is handled for them.</p><h3><a class="nav-anchor" id="Setindex-and-modifying-DNA-sequences-1" href="#Setindex-and-modifying-DNA-sequences-1">Setindex and modifying DNA sequences</a></h3><p>The biological symbol at a given locus in a biological sequence can be set using setindex:</p><pre><code class="language-julia-repl">julia&gt; seq = dna&quot;ACGTTTANAGTNNAGTACC&quot;
19nt DNA Sequence:
ACGTTTANAGTNNAGTACC

julia&gt; seq[5] = DNA_A
DNA_A
</code></pre><p>In addition, many other modifying operations are possible for biological sequences such as <code>push!</code>, <code>pop!</code>, and <code>insert!</code>, which should be familiar to people used to editing arrays.</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.push!" href="#Base.push!"><code>Base.push!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">push!(collection, items...) -&gt; collection</code></pre><p>Insert one or more <code>items</code> at the end of <code>collection</code>.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; push!([1, 2, 3], 4, 5, 6)
6-element Array{Int64,1}:
 1
 2
 3
 4
 5
 6</code></pre><p>Use <a href="#Base.append!"><code>append!</code></a> to add all the elements of another collection to <code>collection</code>. The result of the preceding example is equivalent to <code>append!([1, 2, 3], [4, 5, 6])</code>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/JuliaLang/julia/blob/80516ca20297a67b996caa08c38786332379b6a5/base/array.jl#L828-L848">source</a></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.pop!" href="#Base.pop!"><code>Base.pop!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">pop!(collection) -&gt; item</code></pre><p>Remove an item in <code>collection</code> and return it. If <code>collection</code> is an ordered container, the last item is returned.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; A=[1, 2, 3]
3-element Array{Int64,1}:
 1
 2
 3

julia&gt; pop!(A)
3

julia&gt; A
2-element Array{Int64,1}:
 1
 2

julia&gt; S = Set([1, 2])
Set([2, 1])

julia&gt; pop!(S)
2

julia&gt; S
Set([1])

julia&gt; pop!(Dict(1=&gt;2))
1 =&gt; 2</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/JuliaLang/julia/blob/80516ca20297a67b996caa08c38786332379b6a5/base/array.jl#L1026-L1060">source</a><div><div><pre><code class="language-none">pop!(collection, key[, default])</code></pre><p>Delete and return the mapping for <code>key</code> if it exists in <code>collection</code>, otherwise return <code>default</code>, or throw an error if <code>default</code> is not specified.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; d = Dict(&quot;a&quot;=&gt;1, &quot;b&quot;=&gt;2, &quot;c&quot;=&gt;3);

julia&gt; pop!(d, &quot;a&quot;)
1

julia&gt; pop!(d, &quot;d&quot;)
ERROR: KeyError: key &quot;d&quot; not found
Stacktrace:
[...]

julia&gt; pop!(d, &quot;e&quot;, 4)
4</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/JuliaLang/julia/blob/80516ca20297a67b996caa08c38786332379b6a5/base/dict.jl#L584-L605">source</a><div><div><pre><code class="language-none">pop!(h::BinaryMinMaxHeap) = popmin!(h)</code></pre></div></div><div><div><pre><code class="language-none">pop!(sc, k)</code></pre><p>Deletes the item with key <code>k</code> in SortedDict or SortedSet <code>sc</code> and returns the value that was associated with <code>k</code> in the case of SortedDict or <code>k</code> itself in the case of SortedSet. A <code>KeyError</code> results if <code>k</code> is not in <code>sc</code>. Time: O(<em>c</em> log <em>n</em>)</p></div></div><div><div><pre><code class="language-none">pop!(sc, k)</code></pre><p>Deletes the item with key <code>k</code> in SortedDict or SortedSet <code>sc</code> and returns the value that was associated with <code>k</code> in the case of SortedDict or <code>k</code> itself in the case of SortedSet. A <code>KeyError</code> results if <code>k</code> is not in <code>sc</code>. Time: O(<em>c</em> log <em>n</em>)</p></div></div><div><div><pre><code class="language-none">pop!(ss)</code></pre><p>Deletes the item with first key in SortedSet <code>ss</code> and returns the key. A <code>BoundsError</code> results if <code>ss</code> is empty. Time: O(<em>c</em> log <em>n</em>)</p></div></div><div><div><pre><code class="language-none">pop!(cb)</code></pre><p>Remove the element at the back.</p></div></div><div><div><pre><code class="language-none">pop!(seq::BioSequence)</code></pre><p>Remove the symbol from the end of a biological sequence <code>seq</code> and return it. Returns a variable of <code>eltype(seq)</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.popfirst!" href="#Base.popfirst!"><code>Base.popfirst!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">popfirst!(collection) -&gt; item</code></pre><p>Remove the first <code>item</code> from <code>collection</code>.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; A = [1, 2, 3, 4, 5, 6]
6-element Array{Int64,1}:
 1
 2
 3
 4
 5
 6

julia&gt; popfirst!(A)
1

julia&gt; A
5-element Array{Int64,1}:
 2
 3
 4
 5
 6</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/JuliaLang/julia/blob/80516ca20297a67b996caa08c38786332379b6a5/base/array.jl#L1094-L1121">source</a><div><div><pre><code class="language-none">popfirst!(cb)</code></pre><p>Remove the first item (at the back) from CircularBuffer.</p></div></div><div><div><pre><code class="language-none">popfirst!(seq)</code></pre><p>Remove the symbol from the beginning of a biological sequence <code>seq</code> and return it. Returns a variable of <code>eltype(seq)</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.pushfirst!" href="#Base.pushfirst!"><code>Base.pushfirst!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">pushfirst!(collection, items...) -&gt; collection</code></pre><p>Insert one or more <code>items</code> at the beginning of <code>collection</code>.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; pushfirst!([1, 2, 3, 4], 5, 6)
6-element Array{Int64,1}:
 5
 6
 1
 2
 3
 4</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/JuliaLang/julia/blob/80516ca20297a67b996caa08c38786332379b6a5/base/array.jl#L1070-L1086">source</a><div><div><pre><code class="language-none">pushfirst!(cb, data)</code></pre><p>Insert one or more items at the beginning of CircularBuffer  and overwrite back if full.</p></div></div><div><div><pre><code class="language-none">pushfirst!(seq, x)</code></pre><p>Insert a biological symbol <code>x</code> at the beginning of a biological sequence <code>seq</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.insert!" href="#Base.insert!"><code>Base.insert!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">insert!(a::Vector, index::Integer, item)</code></pre><p>Insert an <code>item</code> into <code>a</code> at the given <code>index</code>. <code>index</code> is the index of <code>item</code> in the resulting <code>a</code>.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; insert!([6, 5, 4, 2, 1], 4, 3)
6-element Array{Int64,1}:
 6
 5
 4
 3
 2
 1</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/JuliaLang/julia/blob/80516ca20297a67b996caa08c38786332379b6a5/base/array.jl#L1131-L1148">source</a><div><div><pre><code class="language-none">insert!(sc, k)</code></pre><p>Argument <code>sc</code> is a SortedDict or SortedMultiDict, <code>k</code> is a key and <code>v</code> is the corresponding value. This inserts the <code>(k,v)</code> pair into the container. If the key is already present in a SortedDict, this overwrites the old value. In the case of SortedMultiDict, no overwriting takes place (since SortedMultiDict allows the same key to associate with multiple values). In the case of SortedDict, the return value is a pair whose first entry is boolean and indicates whether the insertion was new (i.e., the key was not previously present) and the second entry is the semitoken of the new entry. In the case of SortedMultiDict, a semitoken is returned (but no boolean). Time: O(<em>c</em> log <em>n</em>)</p></div></div><div><div><pre><code class="language-none">insert!(sc, k)</code></pre><p>Argument <code>sc</code> is a SortedDict or SortedMultiDict, <code>k</code> is a key and <code>v</code> is the corresponding value. This inserts the <code>(k,v)</code> pair into the container. If the key is already present in a SortedDict, this overwrites the old value. In the case of SortedMultiDict, no overwriting takes place (since SortedMultiDict allows the same key to associate with multiple values). In the case of SortedDict, the return value is a pair whose first entry is boolean and indicates whether the insertion was new (i.e., the key was not previously present) and the second entry is the semitoken of the new entry. In the case of SortedMultiDict, a semitoken is returned (but no boolean). Time: O(<em>c</em> log <em>n</em>)</p></div></div><div><div><pre><code class="language-none">insert!(sc, k)</code></pre><p>Argument <code>sc</code> is a SortedSet and <code>k</code> is a key. This inserts the key into the container. If the key is already present, this overwrites the old value. (This is not necessarily a no-op; see below for remarks about the customizing the sort order.) The return value is a pair whose first entry is boolean and indicates whether the insertion was new (i.e., the key was not previously present) and the second entry is the semitoken of the new entry. Time: O(<em>c</em> log <em>n</em>)</p></div></div><div><div><pre><code class="language-none">insert!(seq, i, x)</code></pre><p>Insert a biological symbol <code>x</code> into a biological sequence <code>seq</code>, at the given index <code>i</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.deleteat!-Tuple{BioSequence,Integer}" href="#Base.deleteat!-Tuple{BioSequence,Integer}"><code>Base.deleteat!</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-none">deleteat!(seq::BioSequence, i::Integer)</code></pre><p>Delete a biological symbol at a single position <code>i</code> in a biological sequence <code>seq</code>.</p><p>Modifies the input sequence.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.append!" href="#Base.append!"><code>Base.append!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">append!(collection, collection2) -&gt; collection.</code></pre><p>Add the elements of <code>collection2</code> to the end of <code>collection</code>.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; append!([1],[2,3])
3-element Array{Int64,1}:
 1
 2
 3

julia&gt; append!([1, 2, 3], [4, 5, 6])
6-element Array{Int64,1}:
 1
 2
 3
 4
 5
 6</code></pre><p>Use <a href="#Base.push!"><code>push!</code></a> to add individual items to <code>collection</code> which are not already themselves in another collection. The result of the preceding example is equivalent to <code>push!([1, 2, 3], 4, 5, 6)</code>.</p></div></div><a class="source-link" target="_blank" href="https://github.com/JuliaLang/julia/blob/80516ca20297a67b996caa08c38786332379b6a5/base/array.jl#L865-L891">source</a><div><div><pre><code class="language-none">append!(cb, datavec)</code></pre><p>Push at most last <code>capacity</code> items.</p></div></div><div><div><p>Write part of a byte array.</p></div></div><div><div><pre><code class="language-none">append!(seq, other)</code></pre><p>Add a biological sequence <code>other</code> onto the end of biological sequence <code>seq</code>. Modifies and returns <code>seq</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.resize!" href="#Base.resize!"><code>Base.resize!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">resize!(a::Vector, n::Integer) -&gt; Vector</code></pre><p>Resize <code>a</code> to contain <code>n</code> elements. If <code>n</code> is smaller than the current collection length, the first <code>n</code> elements will be retained. If <code>n</code> is larger, the new elements are not guaranteed to be initialized.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; resize!([6, 5, 4, 3, 2, 1], 3)
3-element Array{Int64,1}:
 6
 5
 4

julia&gt; a = resize!([6, 5, 4, 3, 2, 1], 8);

julia&gt; length(a)
8

julia&gt; a[1:6]
6-element Array{Int64,1}:
 6
 5
 4
 3
 2
 1</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/JuliaLang/julia/blob/80516ca20297a67b996caa08c38786332379b6a5/base/array.jl#L971-L1000">source</a><div><div><pre><code class="language-none">resize!(seq, size)</code></pre><p>Resize a biological sequence <code>seq</code>, to a given <code>size</code>.</p></div></div></section><p>Here are some examples:</p><pre><code class="language-julia-repl">julia&gt; seq = dna&quot;ACG&quot;
3nt DNA Sequence:
ACG

julia&gt; push!(seq, DNA_T)
4nt DNA Sequence:
ACGT

julia&gt; append!(seq, dna&quot;AT&quot;)
6nt DNA Sequence:
ACGTAT

julia&gt; deleteat!(seq, 2)
5nt DNA Sequence:
AGTAT

julia&gt; deleteat!(seq, 2:3)
3nt DNA Sequence:
AAT
</code></pre><h3><a class="nav-anchor" id="Additional-transformations-1" href="#Additional-transformations-1">Additional transformations</a></h3><p>In addition to these basic modifying functions, other sequence transformations which are common in bioinformatics are also provided.</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.reverse!-Tuple{BioSequence}" href="#Base.reverse!-Tuple{BioSequence}"><code>Base.reverse!</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-none">reverse!(seq)</code></pre><p>Reverse a biological sequence <code>seq</code> in place.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.complement!" href="#BioSequences.complement!"><code>BioSequences.complement!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">complement!(seq)</code></pre><p>Make a complement sequence of <code>seq</code> in place.</p></div></div><div><div><pre><code class="language-none">complement!(seq)</code></pre><p>Transform <code>seq</code> into it&#39;s complement.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.reverse_complement!" href="#BioSequences.reverse_complement!"><code>BioSequences.reverse_complement!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">reverse_complement!(seq)</code></pre><p>Make a reversed complement sequence of <code>seq</code> in place.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.ungap!" href="#BioSequences.ungap!"><code>BioSequences.ungap!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Remove gap characters from a sequence. Modifies the input sequence.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Base.empty!" href="#Base.empty!"><code>Base.empty!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">empty!(collection) -&gt; collection</code></pre><p>Remove all elements from a <code>collection</code>.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl">julia&gt; A = Dict(&quot;a&quot; =&gt; 1, &quot;b&quot; =&gt; 2)
Dict{String,Int64} with 2 entries:
  &quot;b&quot; =&gt; 2
  &quot;a&quot; =&gt; 1

julia&gt; empty!(A);

julia&gt; A
Dict{String,Int64} with 0 entries</code></pre></div></div><a class="source-link" target="_blank" href="https://github.com/JuliaLang/julia/blob/80516ca20297a67b996caa08c38786332379b6a5/base/dict.jl#L246-L263">source</a><div><div><pre><code class="language-none">empty!(cb)</code></pre><p>Reset the buffer.</p></div></div><div><div><pre><code class="language-none">empty!(seq)</code></pre><p>Completely empty a biological sequence <code>seq</code> of nucleotides.</p></div></div></section><p>Some examples:</p><pre><code class="language-julia-repl">julia&gt; seq = dna&quot;ACGTAT&quot;
6nt DNA Sequence:
ACGTAT

julia&gt; reverse!(seq)
6nt DNA Sequence:
TATGCA

julia&gt; complement!(seq)
6nt DNA Sequence:
ATACGT

julia&gt; reverse_complement!(seq)
6nt DNA Sequence:
ACGTAT
</code></pre><p>Many of these methods also have a version which makes a copy of the input sequence, so you get a modified copy, and don&#39;t alter the original sequence. Such methods are named the same, but without the exclamation mark. E.g. <code>reverse</code> instead of <code>reverse!</code>, and <code>ungap</code> instead of <code>ungap!</code>.  </p><h4><a class="nav-anchor" id="Translation-1" href="#Translation-1">Translation</a></h4><p>Translation is a slightly more complex transformation for RNA Sequences and so we describe it here in more detail.</p><p>The <a href="#BioSequences.translate"><code>translate</code></a> funtion translates a sequence of codons in a RNA sequence to a amino acid sequence besed on a genetic code mapping. The <code>BioSequences</code> module contains all NCBI defined genetic codes and they are registered in <a href="#BioSequences.ncbi_trans_table"><code>ncbi_trans_table</code></a>. DNA sequences are automatically converted to RNA before being translated.</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.translate" href="#BioSequences.translate"><code>BioSequences.translate</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">translate(seq, code=standard_genetic_code, allow_ambiguous_codons=true, convert_start_codon=false)</code></pre><p>Translate an <code>RNASequence</code> or a <code>DNASequence</code> to an <code>AminoAcidSequence</code>.</p><p>Translation uses genetic code <code>code</code> to map codons to amino acids. See <code>ncbi_trans_table</code> for available genetic codes. If codons in the given sequence cannot determine a unique amino acid, they will be translated to <code>AA_X</code> if <code>allow_ambiguous_codons</code> is <code>true</code> and otherwise result in an error. For organisms that utilize alternative start codons, one can set <code>alternative_start=true</code>, in which case the first codon will always be converted to a methionine.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.ncbi_trans_table" href="#BioSequences.ncbi_trans_table"><code>BioSequences.ncbi_trans_table</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>Genetic code list of NCBI.</p><p>The standard genetic code is <code>ncbi_trans_table[1]</code> and others can be shown by <code>show(ncbi_trans_table)</code>. For more details, consult the next link: http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes.</p></div></div></section><pre><code class="language-julia-repl">julia&gt; ncbi_trans_table
Translation Tables:
  1. The Standard Code (standard_genetic_code)
  2. The Vertebrate Mitochondrial Code (vertebrate_mitochondrial_genetic_code)
  3. The Yeast Mitochondrial Code (yeast_mitochondrial_genetic_code)
  4. The Mold, Protozoan, and Coelenterate Mitochondrial Code and the Mycoplasma/Spiroplasma Code (mold_mitochondrial_genetic_code)
  5. The Invertebrate Mitochondrial Code (invertebrate_mitochondrial_genetic_code)
  6. The Ciliate, Dasycladacean and Hexamita Nuclear Code (ciliate_nuclear_genetic_code)
  9. The Echinoderm and Flatworm Mitochondrial Code (echinoderm_mitochondrial_genetic_code)
 10. The Euplotid Nuclear Code (euplotid_nuclear_genetic_code)
 11. The Bacterial, Archaeal and Plant Plastid Code (bacterial_plastid_genetic_code)
 12. The Alternative Yeast Nuclear Code (alternative_yeast_nuclear_genetic_code)
 13. The Ascidian Mitochondrial Code (ascidian_mitochondrial_genetic_code)
 14. The Alternative Flatworm Mitochondrial Code (alternative_flatworm_mitochondrial_genetic_code)
 16. Chlorophycean Mitochondrial Code (chlorophycean_mitochondrial_genetic_code)
 21. Trematode Mitochondrial Code (trematode_mitochondrial_genetic_code)
 22. Scenedesmus obliquus Mitochondrial Code (scenedesmus_obliquus_mitochondrial_genetic_code)
 23. Thraustochytrium Mitochondrial Code (thraustochytrium_mitochondrial_genetic_code)
 24. Pterobranchia Mitochondrial Code (pterobrachia_mitochondrial_genetic_code)
 25. Candidate Division SR1 and Gracilibacteria Code (candidate_division_sr1_genetic_code)
</code></pre><p><a href="http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes">http://www.ncbi.nlm.nih.gov/Taxonomy/taxonomyhome.html/index.cgi?chapter=cgencodes</a></p><h2><a class="nav-anchor" id="Site-counting-1" href="#Site-counting-1">Site counting</a></h2><p>BioSequences extends the <code>Base.count</code> method to provide some useful utilities for counting the number of sites in biological sequences.</p><h3><a class="nav-anchor" id="Site-types-1" href="#Site-types-1">Site types</a></h3><p>Different types of site can be counted. Each of these types is a concrete subtype of the abstract type <code>Site</code>:</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.Certain" href="#BioSequences.Certain"><code>BioSequences.Certain</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>A <code>Certain</code> site describes a site where both of two aligned sites are not an ambiguity symbol or a gap.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.Gap" href="#BioSequences.Gap"><code>BioSequences.Gap</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>An <code>Gap</code> site describes a site where either of two aligned sites are a gap symbol &#39;-&#39;.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.Ambiguous" href="#BioSequences.Ambiguous"><code>BioSequences.Ambiguous</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>An <code>Ambiguous</code> site describes a site where either of two aligned sites are an ambiguity symbol.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.Match" href="#BioSequences.Match"><code>BioSequences.Match</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>A <code>Match</code> site describes a site where two aligned nucleotides are the same biological symbol.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.Mismatch" href="#BioSequences.Mismatch"><code>BioSequences.Mismatch</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>A <code>Mismatch</code> site describes a site where two aligned nucleotides are not the same biological symbol.</p></div></div></section><h3><a class="nav-anchor" id="Base.count-methods-1" href="#Base.count-methods-1"><code>Base.count</code> methods</a></h3><p>The count method can be used with two sequences and a concrete subtype of <code>Site</code>:</p><pre><code class="language-julia-repl">julia&gt; count(Match, dna&quot;ATCGATCG&quot;, dna&quot;AAGGTTCG&quot;)
5</code></pre><p>By providing a <code>window</code> and <code>step</code> size, counting can be done from within a sliding window:</p><pre><code class="language-julia-repl">julia&gt; count(Match, dna&quot;ATCGATCG&quot;, dna&quot;AAGGTTCG&quot;, 3, 1)
6-element Array{IntervalTrees.IntervalValue{Int64,Int64},1}:
 IntervalTrees.IntervalValue{Int64,Int64}
(1,3) =&gt; 1
 IntervalTrees.IntervalValue{Int64,Int64}
(2,4) =&gt; 1
 IntervalTrees.IntervalValue{Int64,Int64}
(3,5) =&gt; 1
 IntervalTrees.IntervalValue{Int64,Int64}
(4,6) =&gt; 2
 IntervalTrees.IntervalValue{Int64,Int64}
(5,7) =&gt; 2
 IntervalTrees.IntervalValue{Int64,Int64}
(6,8) =&gt; 3</code></pre><h3><a class="nav-anchor" id="The-pairwise_count-function-1" href="#The-pairwise_count-function-1">The <code>pairwise_count</code> function</a></h3><p>Counting can also be done on a set of sequences in a pairwise manner with the <code>count_pairwise</code> function:</p><pre><code class="language-julia-repl">julia&gt; count_pairwise(Match, dna&quot;ATCGCCA-&quot;, dna&quot;ATCGCCTA&quot;, dna&quot;ATCGCCT-&quot;, dna&quot;GTCGCCTA&quot;)
4×4 Array{Int64,2}:
 0  6  7  5
 6  0  7  7
 7  7  0  6
 5  7  6  0</code></pre><h2><a class="nav-anchor" id="Iteration-1" href="#Iteration-1">Iteration</a></h2><p>Sequences also work as iterators over symbols:</p><pre><code class="language-julia-repl">julia&gt; n = 0
0

julia&gt; for nt in dna&quot;ATNGNNT&quot;
           if nt == DNA_N
               global n += 1
           end
       end

julia&gt; n
3
</code></pre><h2><a class="nav-anchor" id="Using-a-more-compact-sequence-representation-1" href="#Using-a-more-compact-sequence-representation-1">Using a more compact sequence representation</a></h2><p>As we saw above, DNA and RNA sequences can store any ambiguous nucleotides like &#39;N&#39;.  If you are sure that nucleotide sequences store unambiguous nucleotides only, you can save the memory space of sequences. <code>DNAAlphabet{2}</code> is an alphabet that uses two bits per base and limits to only unambiguous nucleotide symbols (ACGT in DNA and ACGU in RNA). To create a sequence of this alphabet, you need to explicitly pass <code>DNAAlphabet{2}</code> to <code>BioSequence</code> as its type parameter:</p><pre><code class="language-julia-repl">julia&gt; seq = BioSequence{DNAAlphabet{2}}(&quot;ACGT&quot;)
4nt DNA Sequence:
ACGT
</code></pre><p>Recall that <code>DNASequence</code> is a type alias of <code>BioSequence{DNAAlphabet{4}}</code>, which uses four bits per base. That is, <code>BioSequence{DNAAlphabet{2}}</code> saves half of memory footprint compared to <code>BioSequence{DNAAlphabet{4}}</code>. If you need to handle reference genomes that are composed of five nucleotides, ACGTN, consider to use the <code>ReferenceSequence</code> type described in the <a href="../refseq/#Reference-sequences-1">Reference sequences</a> section.</p><h2><a class="nav-anchor" id="Defining-a-new-alphabet-1" href="#Defining-a-new-alphabet-1">Defining a new alphabet</a></h2><p>The alphabet type parameter <code>A</code> of <code>BioSequence{A}</code> enables a user to extend functionality of <code>BioSequence</code> with minimum effort. As an example, definition of a new alphabet type representing a sequence of boolean values is shown below:</p><pre><code class="language-julia-repl">julia&gt; struct BoolAlphabet &lt;: Alphabet end

julia&gt; BioSequences.bitsof(::Type{BoolAlphabet}) = 1

julia&gt; BioSequences.eltype(::Type{BoolAlphabet}) = Bool

julia&gt; BioSequences.alphabet(::Type{BoolAlphabet}) = false:true

julia&gt; function BioSequences.encode(::Type{BoolAlphabet}, x::Bool)
           return UInt64(ifelse(x, 0x01, 0x00))
       end

julia&gt; function BioSequences.decode(::Type{BoolAlphabet}, x::UInt64)
           if x &gt; 0x01
               throw(BioSequences.DecodeError(BoolAlphabet, x))
           end
           return ifelse(x == 0x00, false, true)
       end
</code></pre><footer><hr/><a class="previous" href="../sequences/"><span class="direction">Previous</span><span class="title">Overview</span></a><a class="next" href="../refseq/"><span class="direction">Next</span><span class="title">Reference Sequences</span></a></footer></article></body></html>
