<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Overview Â· BioSequences.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="BioSequences.jl logo"/></a><h1>BioSequences.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><a class="toctext" href="../../symbols/">Biological Symbols</a></li><li><span class="toctext">Sequence types</span><ul><li class="current"><a class="toctext" href>Overview</a><ul class="internal"><li><a class="toctext" href="#Introduction-to-the-sequence-data-types-1">Introduction to the sequence data-types</a></li></ul></li><li><a class="toctext" href="../bioseq/">BioSequence</a></li><li><a class="toctext" href="../refseq/">Reference Sequences</a></li><li><a class="toctext" href="../kmer/">Nucleic acid k-mers</a></li></ul></li><li><span class="toctext">IO</span><ul><li><a class="toctext" href="../../io/fasta/">FASTA formatted files</a></li><li><a class="toctext" href="../../io/fastq/">FASTQ formatted files</a></li><li><a class="toctext" href="../../io/twobit/">2bit formatted files</a></li></ul></li><li><a class="toctext" href="../../search/">Searching</a></li><li><a class="toctext" href="../../composition/">Sequence Composition</a></li><li><a class="toctext" href="../../demultiplexer/">Demultiplexing</a></li><li><a class="toctext" href="../../contributing/">Contributing</a></li></ul></nav><article id="docs"><header><nav><ul><li>Sequence types</li><li><a href>Overview</a></li></ul></nav><hr/><div id="topbar"><span>Overview</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Biological-sequences-1" href="#Biological-sequences-1">Biological sequences</a></h1><p>The <code>BioSequences</code> module provides representations and tools for manipulating nucleotide and amino acid sequences.</p><h2><a class="nav-anchor" id="Introduction-to-the-sequence-data-types-1" href="#Introduction-to-the-sequence-data-types-1">Introduction to the sequence data-types</a></h2><p>Sequences in BioSequences.jl are more strictly typed than in many other libraries; elements in a sequence are typed as biological symbol instead of character or byte. They are special purpose types rather than simply strings and hence offer additional functionality that naive string types don&#39;t have. Though this strictness sacrifices some convenience, it also means you can always rely on a DNA sequence type to store DNA and nothing but DNA, without having to check, or deal with lowercase versus uppercase and so on. Strict separation of sequence types also means we are free to choose the most efficient representation. DNA and RNA sequences are encoded using either four bits per base (which is the default), or two bits per base. This makes them memory efficient and allows us to speed up many common operations and transformations, like nucleotide composition, reverse complement, and <em>k</em>-mer enumeration.</p><p>The <code>BioSequences</code> provides three different sequence types: <code>BioSequence</code>, <code>Kmer</code> and <code>ReferenceSequence</code>. Each of these types is a subtype of an abstract type called <code>Sequence</code> and supports various string-like operations such as random access and iteration. Different sequence types have different features. In most situations, <code>BioSequence</code> type will do and is used as the default representation. But sometimes other types are much more preferable in terms of memory efficiency and computation performance.  Here is the summary table of these three types:</p><table><tr><th>Type</th><th>Description</th><th>Element type</th><th>Mutability</th><th>Allocation</th></tr><tr><td><code>BioSequence{A&lt;:Alphabet}</code></td><td>general-purpose biological sequences</td><td>DNA, RNA, Amino acids</td><td>mutable</td><td>heap</td></tr><tr><td><code>Kmer{T&lt;:NucleicAcid,k}</code></td><td>specialized for short nucleotide sequences</td><td>DNA, RNA</td><td>immutable</td><td>stack / register</td></tr><tr><td><code>ReferenceSequence</code></td><td>specialized for long reference genomes</td><td>DNA</td><td>immutable</td><td>heap</td></tr></table><p>Details of these different representations are explained in the following sections:</p><ul><li><code>BioSequence</code>: <a href="../bioseq/#General-purpose-sequences-1">General-purpose sequences</a></li><li><code>Kmer</code>: <a href="../kmer/#Nucleic-acid-k-mers-1">Nucleic acid k-mers</a></li><li><code>ReferenceSequence</code>: <a href="../refseq/#Reference-sequences-1">Reference sequences</a></li></ul><footer><hr/><a class="previous" href="../../symbols/"><span class="direction">Previous</span><span class="title">Biological Symbols</span></a><a class="next" href="../bioseq/"><span class="direction">Next</span><span class="title">BioSequence</span></a></footer></article></body></html>
