<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Docstrings · ProteinEnsembles.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>ProteinEnsembles.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../">Readme</a></li><li class="current"><a class="toctext" href>Docstrings</a><ul class="internal"></ul></li></ul></nav><article id="docs"><header><nav><ul><li><a href>Docstrings</a></li></ul></nav><hr/><div id="topbar"><span>Docstrings</span><a class="fa fa-bars" href="#"></a></div></header><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.Atom" href="#ProteinEnsembles.Atom"><code>ProteinEnsembles.Atom</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>An atom from a PDB format file.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.BondedPair" href="#ProteinEnsembles.BondedPair"><code>ProteinEnsembles.BondedPair</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>A pair of atoms with a covalent bond between them in the standard amino acids.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.Constraints" href="#ProteinEnsembles.Constraints"><code>ProteinEnsembles.Constraints</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Information about the lower and upper distance constraints between atoms.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.ModelledEnsemble" href="#ProteinEnsembles.ModelledEnsemble"><code>ProteinEnsembles.ModelledEnsemble</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>A group of modelled structures.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.ModelledStructure" href="#ProteinEnsembles.ModelledStructure"><code>ProteinEnsembles.ModelledStructure</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>A set of coordinates and a score.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.PCA" href="#ProteinEnsembles.PCA"><code>ProteinEnsembles.PCA</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>The results of a principal components analysis, sorted with highest eigenvalue first.</p></div></div></section><pre><code class="language-none">ProteinEnsembles.ProteinEnsembles</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.addstructure!" href="#ProteinEnsembles.addstructure!"><code>ProteinEnsembles.addstructure!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Add a structure to an ensemble.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.align!" href="#ProteinEnsembles.align!"><code>ProteinEnsembles.align!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Align the first set of atomic coordinates to the second using an iterative procedure. At each step atoms with a high structural deviation are removed. If the standard deviation of the structural deviations is greater than <code>cutoff</code>, then structural deviations further than 2 standard deviations from the median are discarded. This is repeated <code>n_cycles</code> times. If <code>calpha</code> is true, only considers the C-alpha atoms.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.alignatoms!" href="#ProteinEnsembles.alignatoms!"><code>ProteinEnsembles.alignatoms!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Align an array of atoms to a set of coordinates or another set of atoms. Input is list of atoms and coords to align to. Changes the atom coords.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.alignensemble!" href="#ProteinEnsembles.alignensemble!"><code>ProteinEnsembles.alignensemble!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Align an ensemble of coordinates to a coordinate set using an iterative procedure for each structure.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.alignensemblesimple!" href="#ProteinEnsembles.alignensemblesimple!"><code>ProteinEnsembles.alignensemblesimple!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Align an ensemble of coordinates to a coordinate set using a single Kabsch rotation for each structure.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.alignsimple!" href="#ProteinEnsembles.alignsimple!"><code>ProteinEnsembles.alignsimple!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Align the first set of atomic coordinates to the second using a single Kabsch rotation. If <code>calpha</code> is true, only considers the C-alpha atoms.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.angles" href="#ProteinEnsembles.angles"><code>ProteinEnsembles.angles</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Calculate the bond angle matrix for a list of atoms. Returns a 2D boolean array of whether bond angles are present.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.areatomssame" href="#ProteinEnsembles.areatomssame"><code>ProteinEnsembles.areatomssame</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Determines if two atoms are the same. Atoms are the same if they have the same <code>atom_name</code>, <code>chain_id</code>, <code>res_n</code> and <code>element</code>. <code>res_name</code> is not considered.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.atomcoords" href="#ProteinEnsembles.atomcoords"><code>ProteinEnsembles.atomcoords</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Get or set the list of coordinates for a list of atoms.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.atomcoords!" href="#ProteinEnsembles.atomcoords!"><code>ProteinEnsembles.atomcoords!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Set the coordinates of a list of atoms from a list of coordinates.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.atomid" href="#ProteinEnsembles.atomid"><code>ProteinEnsembles.atomid</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>A string describing an <code>Atom</code> in the form <code>atom_name</code>/<code>res_name</code>/<code>res_n</code>/<code>chain_id</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.atommap" href="#ProteinEnsembles.atommap"><code>ProteinEnsembles.atommap</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Finds the mapping of the atoms in the first list to those in the second. Returns an array of length the same as the second list where the number is the atom index in the first list. Atoms missing from the first list are given 0.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.bonds" href="#ProteinEnsembles.bonds"><code>ProteinEnsembles.bonds</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Calculate the bonding matrix for a list of atoms. Returns a 2D boolean array of bonding.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.calphaindices" href="#ProteinEnsembles.calphaindices"><code>ProteinEnsembles.calphaindices</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Return as an array the indices in an array of atoms that correspond to C-alpha atoms.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.centroid" href="#ProteinEnsembles.centroid"><code>ProteinEnsembles.centroid</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Find the average structure in an ensemble. The ensemble should be aligned.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.checkfilepath" href="#ProteinEnsembles.checkfilepath"><code>ProteinEnsembles.checkfilepath</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Checks if an output filepath is valid, and throws an error if not.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.clusterligsite" href="#ProteinEnsembles.clusterligsite"><code>ProteinEnsembles.clusterligsite</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Assign pocket numbers to LIGSITEcs pocket<em>r.pdb file points. Arguments are pocket</em>r.pdb file, pocket<em>all.pdb file and new pocket</em>r.pdb file. New file has pocket number in place of residue number. Apears to agree largely but not completely with the LIGSITEcs clustering. Unassigned points given number 0 and should be fixed manually.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.default_colours" href="#ProteinEnsembles.default_colours"><code>ProteinEnsembles.default_colours</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>Dictionary of default colours for plotting.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.defaults" href="#ProteinEnsembles.defaults"><code>ProteinEnsembles.defaults</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>Dictionary of default values for parameters.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.displacements" href="#ProteinEnsembles.displacements"><code>ProteinEnsembles.displacements</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Return the distances between each point for two sets of aligned coordinates.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.distances" href="#ProteinEnsembles.distances"><code>ProteinEnsembles.distances</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Calculate the distance matrix for a list of atoms. Returns a 2D array of distances.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.divalents" href="#ProteinEnsembles.divalents"><code>ProteinEnsembles.divalents</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Calculate the divalent matrix for a list of atoms. Returns a 2D boolean array of whether divalents are present.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.ensemblermsds" href="#ProteinEnsembles.ensemblermsds"><code>ProteinEnsembles.ensemblermsds</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Get the RMSD between each structure in an ensemble and a reference structure.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.errorscore" href="#ProteinEnsembles.errorscore"><code>ProteinEnsembles.errorscore</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Calculate SPE error score.</p></div></div></section><pre><code class="language-none">ProteinEnsembles.eval</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.fillconstraints" href="#ProteinEnsembles.fillconstraints"><code>ProteinEnsembles.fillconstraints</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Use stochastic proximity embedding to generate a structure iteratively. See Agrafiotis et.al., Chapter 14 in Mucherino et. al., 2013 for algorithm details (Algorithm 8). Returns coordinate list.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.findcommonatoms" href="#ProteinEnsembles.findcommonatoms"><code>ProteinEnsembles.findcommonatoms</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Checks two atom arrays and finds common atoms. Returns copies of input atom arrays with only common atoms remaining. The returned arrays are therefore of equal length.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.findimportantpcs" href="#ProteinEnsembles.findimportantpcs"><code>ProteinEnsembles.findimportantpcs</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Takes two sets of projections onto the PCs of a PCA and returns the PC indices ordered such that the first represents the largest difference between the two structures.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.findinteractions" href="#ProteinEnsembles.findinteractions"><code>ProteinEnsembles.findinteractions</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Find the interactions between atoms in a list of atoms. Returns 2D array of interactions by integer label. Interactions the same as De Groot et. al., Structure, 1997, but cis/trans limits not implemented.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.fixchirality!" href="#ProteinEnsembles.fixchirality!"><code>ProteinEnsembles.fixchirality!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Checks the chirality of coordinate set and switches it if it isn&#39;t correct. The coordinates must correspond to the atoms. Inverts each z coordinate in the coordinate array if required.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.fixstring" href="#ProteinEnsembles.fixstring"><code>ProteinEnsembles.fixstring</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Converts a string, character or number to a fixed length string. Be careful - could e.g. cut 0.5 to &quot;0&quot;. Also does not round correctly.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.fluctuations" href="#ProteinEnsembles.fluctuations"><code>ProteinEnsembles.fluctuations</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Gets mean square fluctuation for each atom in an ensemble. Assumes the ensemble is aligned sensibly. Returns a list of mean square fluctuations.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.fractionbetweeninputs" href="#ProteinEnsembles.fractionbetweeninputs"><code>ProteinEnsembles.fractionbetweeninputs</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Fraction of structures in an ensemble that are closer to both input structures than the input structures are to each other.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.generateensemble" href="#ProteinEnsembles.generateensemble"><code>ProteinEnsembles.generateensemble</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Generate sets of coordinates with the correct chirality from distance constraints. discard<em>ratio gives the fraction of extra structures to calculate, then the low-scoring structures are discarded. If `remove</em>mod<code>is true then modulator atoms and coords are removed from the ensemble before it is returned. Returns a</code>ModelledEnsemble`.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.hard_sphere_radius" href="#ProteinEnsembles.hard_sphere_radius"><code>ProteinEnsembles.hard_sphere_radius</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>Dictionary of atomic hard sphere radii in Angstroms.</p></div></div></section><pre><code class="language-none">ProteinEnsembles.include</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.inferelement" href="#ProteinEnsembles.inferelement"><code>ProteinEnsembles.inferelement</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Determines the element of an atom from the atom name. Returns the element as a string.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.interactioninfo" href="#ProteinEnsembles.interactioninfo"><code>ProteinEnsembles.interactioninfo</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Returns a list with interaction types and a list with corresponding distance tolerances. Zero index represents same atom. <code>tolerance_weight</code> weights all tolerances.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.interactions" href="#ProteinEnsembles.interactions"><code>ProteinEnsembles.interactions</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Run the interaction finding pipeline with one or two structures. If one structure is given, returns a <code>Constraints</code> object. If two structures are given, returns three <code>Constraints</code> objects corresponding to the combined constraints, constraints from structure one and constraints from structure two.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.isdbonefour" href="#ProteinEnsembles.isdbonefour"><code>ProteinEnsembles.isdbonefour</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Determines if two atoms are a 1-4 pair around a restricted side chain double bond. Assumes atoms are a 1-4 pair.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.ishbond" href="#ProteinEnsembles.ishbond"><code>ProteinEnsembles.ishbond</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Determines if two atoms are part of a hydrogen bond where the donor-acceptor distance is &lt;= 3.5 Angstroms, the hydrogen-acceptor distance is &lt;= 2.5 Angstroms and the donor-hydrogen-acceptor angle is minimally 90 degrees. Won&#39;t pick up the donor-H pairing, even though it is part of the H bond, as this pairing is covalent. Sulfur not considered.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.ishydrophobic" href="#ProteinEnsembles.ishydrophobic"><code>ProteinEnsembles.ishydrophobic</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Determines if two atoms are closer than the sum of the VDW radii plus extra<em>dist Angstroms. Currently counting C and H atoms only. Applies a crude 5 A distance cutoff for speed, so extra</em>dist values above ~1.5 A will have no effect.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.isloosephipsi" href="#ProteinEnsembles.isloosephipsi"><code>ProteinEnsembles.isloosephipsi</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Determines if two atoms in a 1-4 pair around the phi/psi angle are either in a loop region or one of the residues is glycine. Assumes atoms are a 1-4 pair around the phi/psi angle.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.ismodulator" href="#ProteinEnsembles.ismodulator"><code>ProteinEnsembles.ismodulator</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Determines whether an <code>Atom</code> is a fake modulator atom.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.isomonefour" href="#ProteinEnsembles.isomonefour"><code>ProteinEnsembles.isomonefour</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Determines if two atoms are a 1-4 pair around the omega angle. Assumes atoms are a 1-4 pair.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.isphipsi" href="#ProteinEnsembles.isphipsi"><code>ProteinEnsembles.isphipsi</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Determines if two atoms are a 1-4 pair around the phi/psi angle. Assumes atoms are a 1-4 pair. Glycine H is not polar so should not be present.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.isring" href="#ProteinEnsembles.isring"><code>ProteinEnsembles.isring</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Determines if two atoms are part of the same ring system.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.issaltbridge" href="#ProteinEnsembles.issaltbridge"><code>ProteinEnsembles.issaltbridge</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Determines if two atoms are part of oppositely charged groups that are within 4 Angstroms. His not included.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.issecstr" href="#ProteinEnsembles.issecstr"><code>ProteinEnsembles.issecstr</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Determines if two atoms are backbone atoms that are part of the same helix/strand and not more than 4 residues apart. Does not include 3-10 and pi helices.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.istightphipsi" href="#ProteinEnsembles.istightphipsi"><code>ProteinEnsembles.istightphipsi</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Determines if two atoms in a 1-4 pair around the phi/psi angle are either in the same helix/strand or one of the residues is proline. Assumes atoms are a 1-4 pair around the phi/psi angle.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.kabschalignment" href="#ProteinEnsembles.kabschalignment"><code>ProteinEnsembles.kabschalignment</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Find the transformations to map the first set of coordinates onto the second. See Kabsch algorithm for details. Returns the translation to center the first set, the translation to center the second set and the rotation to map the first set onto the second.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.makedirectories" href="#ProteinEnsembles.makedirectories"><code>ProteinEnsembles.makedirectories</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Make the outer results directory and any inner directories from a list.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.mod_atom_info" href="#ProteinEnsembles.mod_atom_info"><code>ProteinEnsembles.mod_atom_info</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>Dictionary of atomic information for fake modulator atoms.</p></div></div></section><pre><code class="language-none">ProteinEnsembles.newcoords</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.paramfromshell" href="#ProteinEnsembles.paramfromshell"><code>ProteinEnsembles.paramfromshell</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Run auto-parameterisation from command line.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.parampipeline" href="#ProteinEnsembles.parampipeline"><code>ProteinEnsembles.parampipeline</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Run the auto-parameterisation pipeline.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.parsecommandlineparam" href="#ProteinEnsembles.parsecommandlineparam"><code>ProteinEnsembles.parsecommandlineparam</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Read arguments from command line for an ExProSE auto-parameterisation.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.parsecommandlinerun" href="#ProteinEnsembles.parsecommandlinerun"><code>ProteinEnsembles.parsecommandlinerun</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Read arguments from command line for an ExProSE run.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.pcsaddlayer!" href="#ProteinEnsembles.pcsaddlayer!"><code>ProteinEnsembles.pcsaddlayer!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Adds a layer of points to an existing layer for PC plotting.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.perturbensemble" href="#ProteinEnsembles.perturbensemble"><code>ProteinEnsembles.perturbensemble</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Gets ensembles for multiple modulators. Use pocket points to generate additional constraints.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.plotfluctuations" href="#ProteinEnsembles.plotfluctuations"><code>ProteinEnsembles.plotfluctuations</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Plot root mean square fluctuations of an aligned ensemble. Arguments are the output filepath and the array of fluctuations.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.plotpcs" href="#ProteinEnsembles.plotpcs"><code>ProteinEnsembles.plotpcs</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Plot the projections of an ensemble onto the principal components of the motion.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.plottheme" href="#ProteinEnsembles.plottheme"><code>ProteinEnsembles.plottheme</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p><code>Theme</code> for plots.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.pointtheme" href="#ProteinEnsembles.pointtheme"><code>ProteinEnsembles.pointtheme</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p><code>Theme</code> for plot points.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.projectensemble" href="#ProteinEnsembles.projectensemble"><code>ProteinEnsembles.projectensemble</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Project the structures in an ensemble onto the eigenvectors of a PCA to get the principal coordinates. Non-zero eigenvalues only are returned. Assumes the ensemble is aligned and corresponds exactly to the PCA.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.projectpdbfile" href="#ProteinEnsembles.projectpdbfile"><code>ProteinEnsembles.projectpdbfile</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Projects the structure in a PDB file onto the eigenvectors of a PCA. The residue numbering between the PDB files must be consistent but missing/extra residues are allowed. Reads the PDB file, maps the atoms to the ref atoms, aligns the atoms to the ref atoms then gets the projections. Arguments are PDB filepath, reference atoms to get atom map, coords to align to and the PCA object.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.projectpdbfiles" href="#ProteinEnsembles.projectpdbfiles"><code>ProteinEnsembles.projectpdbfiles</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Wrapper function for projectpdbfile to project multiple PDBs and write the output.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.projectstructure" href="#ProteinEnsembles.projectstructure"><code>ProteinEnsembles.projectstructure</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Project a list of atoms onto the eigenvectors of a PCA. Atom map is the mapping of the input atom indices to the reference atoms. It is not required if the atoms correspond to the PCA. <code>calpha_inds</code> is only required if atom_map is non-trivial and is a list of C-alpha indices. 0 means the atom is not in the reference atoms. Assumes the structure is aligned.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.protein_bonds" href="#ProteinEnsembles.protein_bonds"><code>ProteinEnsembles.protein_bonds</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>A list of bonded pairs representing the covalent bonds in a standard protein.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.readdssp" href="#ProteinEnsembles.readdssp"><code>ProteinEnsembles.readdssp</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Reads a DSSP file. If atoms are provided, creates additional &#39;-&#39; entries for residues not present in the DSSP file. Returns a dictionary where key is residue number and chain, and value is secondary structure.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.readensemble" href="#ProteinEnsembles.readensemble"><code>ProteinEnsembles.readensemble</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Read an ensemble of structures and return a ModelledEnsemble.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.readligsite" href="#ProteinEnsembles.readligsite"><code>ProteinEnsembles.readligsite</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Read a LIGSITEcs PDB output file. Returns the pocket centre coordinate list and list of pocket volumes.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.readpdb" href="#ProteinEnsembles.readpdb"><code>ProteinEnsembles.readpdb</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Parse a PDB file. Returns a list of atoms. hetatm determines whether HETATM records are read in.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.readpdblines" href="#ProteinEnsembles.readpdblines"><code>ProteinEnsembles.readpdblines</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Read the ATOM lines from a PDB file into an array.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.readpocketpoints" href="#ProteinEnsembles.readpocketpoints"><code>ProteinEnsembles.readpocketpoints</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Read a custom pocket point file in PDB format where the residue number is the pocket number. This file can be generated using a script from the LIGSITEcs pocket<em>r.pdb and pocket</em>all.pdb files. Returns a dictionary where key is pocket number and value is point coords.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.removemodulator!" href="#ProteinEnsembles.removemodulator!"><code>ProteinEnsembles.removemodulator!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Takes a <code>ModelledEnsemble</code> and removes the fake modulator atoms and coords.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.repeatpocketpoints" href="#ProteinEnsembles.repeatpocketpoints"><code>ProteinEnsembles.repeatpocketpoints</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Repeat pocket point to get a certain number of points. Points are repeated as long as whole copies can be made, then the remainder are chosen randomly. Returns coordinate array.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.rmsd" href="#ProteinEnsembles.rmsd"><code>ProteinEnsembles.rmsd</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Get the RMSD between two identical structures or two sets of coordinates. Assumes the structures are aligned. If <code>calpha</code> is true, only considers the C-alpha atoms.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.runanalysis" href="#ProteinEnsembles.runanalysis"><code>ProteinEnsembles.runanalysis</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Run the analysis pipeline and write output files.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.runfromshell" href="#ProteinEnsembles.runfromshell"><code>ProteinEnsembles.runfromshell</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Run pipeline from command line.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.runpipeline" href="#ProteinEnsembles.runpipeline"><code>ProteinEnsembles.runpipeline</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Wrapper function to run the whole pipeline.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.selfalignensemble!" href="#ProteinEnsembles.selfalignensemble!"><code>ProteinEnsembles.selfalignensemble!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Align an ensemble to itself. See Bakan and Bahar, PNAS, 2009 for algorithm. The single Kabsch alignment is used during iteration, then the iterative alignment is done at the end. Returns the structure the final alignment was on, not the ensemble average (but they will be close). The cutoff parameter determines the RMSD between two average structures below which iteration is terminated.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.spaceatomname" href="#ProteinEnsembles.spaceatomname"><code>ProteinEnsembles.spaceatomname</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Space an atom name correctly in the four-character limit. The element is in position 2. Returns a string of length 4.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.tmscore" href="#ProteinEnsembles.tmscore"><code>ProteinEnsembles.tmscore</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>TM-score of model and reference, given as filepaths. Uses external command line program - see https://zhanglab.ccmb.med.umich.edu/TM-score.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.tmscorepathvalid" href="#ProteinEnsembles.tmscorepathvalid"><code>ProteinEnsembles.tmscorepathvalid</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Check a command runs TMscore.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.trimbyscore!" href="#ProteinEnsembles.trimbyscore!"><code>ProteinEnsembles.trimbyscore!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Retains only the n<em>to</em>keep top scoring coordinate sets in an ensemble.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.vdw_radius" href="#ProteinEnsembles.vdw_radius"><code>ProteinEnsembles.vdw_radius</code></a> — <span class="docstring-category">Constant</span>.</div><div><div><p>Dictionary of atomic VDW radii in Angstroms.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.writeclusterpoints" href="#ProteinEnsembles.writeclusterpoints"><code>ProteinEnsembles.writeclusterpoints</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Write output file with assignment in residue number column</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.writeensemble" href="#ProteinEnsembles.writeensemble"><code>ProteinEnsembles.writeensemble</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Write an ensemble out as separate PDB files.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.writeensemblescores" href="#ProteinEnsembles.writeensemblescores"><code>ProteinEnsembles.writeensemblescores</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Write out the scores associated with the structures in an ensemble.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.writefloatarray" href="#ProteinEnsembles.writefloatarray"><code>ProteinEnsembles.writefloatarray</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Write an array of floats to an output file.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.writeintarray" href="#ProteinEnsembles.writeintarray"><code>ProteinEnsembles.writeintarray</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Write an array of integers to an output file.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.writepcview" href="#ProteinEnsembles.writepcview"><code>ProteinEnsembles.writepcview</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Write a PyMol script to view the atom motions of a principal component.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.writepcviews" href="#ProteinEnsembles.writepcviews"><code>ProteinEnsembles.writepcviews</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Wrapper function for writepcview that produces visualisation scripts for the top <code>n_to_write</code> PCs.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.writepdb" href="#ProteinEnsembles.writepdb"><code>ProteinEnsembles.writepdb</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Write a list of atoms as a PDB file. The atoms are not reordered so it is assumed they are in a sensible order, e.g. the order they were read in. Only ATOM records are written; no TER labels. Could write a brief header for generated models.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.writeprojections" href="#ProteinEnsembles.writeprojections"><code>ProteinEnsembles.writeprojections</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Write out the projections of structures onto the eigenvectors of a PCA. pcs<em>to</em>write is a list of the PCs to write, e.g. [1, 2]. All PCs written by default.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="ProteinEnsembles.writestringarray" href="#ProteinEnsembles.writestringarray"><code>ProteinEnsembles.writestringarray</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Write an array of strings to an output file.</p></div></div></section><footer><hr/><a class="previous" href="../"><span class="direction">Previous</span><span class="title">Readme</span></a></footer></article></body></html>
