<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>VCF and BCF formatted files Â· GeneticVariation.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../../assets/documenter.js"></script><script src="../../../siteinfo.js"></script><script src="../../../../versions.js"></script><link href="../../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../../index.html"><img class="logo" src="../../../assets/logo.png" alt="GeneticVariation.jl logo"/></a><h1>GeneticVariation.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../../">Home</a></li><li><span class="toctext">Manual</span><ul><li><span class="toctext">I/O</span><ul><li class="current"><a class="toctext" href>VCF and BCF formatted files</a><ul class="internal"></ul></li></ul></li><li><a class="toctext" href="../../site_counting/">Mutation Counting</a></li><li><a class="toctext" href="../../diversity/">Genetic Diversity</a></li></ul></li></ul></nav><article id="docs"><header><nav><ul><li>Manual</li><li>I/O</li><li><a href>VCF and BCF formatted files</a></li></ul></nav><hr/><div id="topbar"><span>VCF and BCF formatted files</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="VCF-and-BCF-Formatted-files-1" href="#VCF-and-BCF-Formatted-files-1">VCF and BCF Formatted files</a></h1><p>VCF is a text-based file format for representing genetic polymorphism.</p><p>VCF files can be read using <code>VCFReader</code>, respectively:</p><pre><code class="language-julia">reader = VCF.Reader(open(&quot;example.vcf&quot;, &quot;r&quot;))
for record in reader
    # do something
end
close(reader)</code></pre><p>A reader first reads the header section of a file and creates a <code>VCF.Header</code> object. The <code>header</code> function is provided to access the header object of a reader:</p><pre><code class="language-jlcon">julia&gt; header(reader)
GeneticVariation.VCF.Header:
  metainfo tags: fileformat fileDate source reference contig phasing INFO FILTER FORMAT
     sample IDs: NA00001 NA00002 NA00003

julia&gt; findall(header(reader), &quot;FORMAT&quot;)
4-element Array{GeneticVariation.VCF.MetaInfo,1}:
 GeneticVariation.VCF.MetaInfo:
    tag: FORMAT
  value: ID=&quot;GT&quot; Number=&quot;1&quot; Type=&quot;String&quot; Description=&quot;Genotype&quot;          
 GeneticVariation.VCF.MetaInfo:
    tag: FORMAT
  value: ID=&quot;GQ&quot; Number=&quot;1&quot; Type=&quot;Integer&quot; Description=&quot;Genotype Quality&quot;
 GeneticVariation.VCF.MetaInfo:
    tag: FORMAT
  value: ID=&quot;DP&quot; Number=&quot;1&quot; Type=&quot;Integer&quot; Description=&quot;Read Depth&quot;       
 GeneticVariation.VCF.MetaInfo:
    tag: FORMAT
  value: ID=&quot;HQ&quot; Number=&quot;2&quot; Type=&quot;Integer&quot; Description=&quot;Haplotype Quality&quot;</code></pre><p><code>VCF.MetaInfo</code> variables in the header support the following accessors:</p><table><tr><th>Accessor</th><th>Description</th></tr><tr><td><code>metainfotag</code></td><td>tag string</td></tr><tr><td><code>metainfoval</code></td><td>value string</td></tr><tr><td><code>keys</code></td><td>keys of fields between &#39;&lt;&#39; and &#39;&gt;&#39;</td></tr><tr><td><code>values</code></td><td>values of fields between &#39;&lt;&#39; and &#39;&gt;&#39;</td></tr><tr><td><code>[&lt;key&gt;]</code></td><td>value of a field with <code>key</code></td></tr></table><pre><code class="language-jlcon">julia&gt; metainfo = VCF.MetaInfo(&quot;##FORMAT=&lt;ID=GT,Number=1,Type=String,Description=\&quot;Genotype\&quot;&gt;&quot;)
GeneticVariation.VCF.MetaInfo:
    tag: FORMAT
  value: ID=&quot;GT&quot; Number=&quot;1&quot; Type=&quot;String&quot; Description=&quot;Genotype&quot;

julia&gt; metainfotag(metainfo)
&quot;FORMAT&quot;

julia&gt; metainfoval(metainfo)
&quot;&lt;ID=GT,Number=1,Type=String,Description=\&quot;Genotype\&quot;&gt;&quot;

julia&gt; keys(metainfo)
4-element Array{String,1}:
 &quot;ID&quot;         
 &quot;Number&quot;     
 &quot;Type&quot;       
 &quot;Description&quot;

julia&gt; metainfo[&quot;ID&quot;]
&quot;GT&quot;
</code></pre><p><code>VCF.Record</code> and <code>BCF.Record</code> variables support the following accessor functions (see the docstring of each accessor for the details):</p><table><tr><th>Accessor</th><th>Description</th></tr><tr><td><code>chrom</code></td><td>chromosome name</td></tr><tr><td><code>pos</code></td><td>reference position</td></tr><tr><td><code>id</code></td><td>unique identifiers</td></tr><tr><td><code>ref</code></td><td>reference bases</td></tr><tr><td><code>alt</code></td><td>alternate bases</td></tr><tr><td><code>qual</code></td><td>Phred-scaled quality score</td></tr><tr><td><code>filter</code></td><td>filter status</td></tr><tr><td><code>info</code></td><td>additional information</td></tr><tr><td><code>infokeys</code></td><td>keys of additional information</td></tr><tr><td><code>format</code></td><td>genotype format</td></tr><tr><td><code>genotype</code></td><td>genotype information</td></tr></table><pre><code class="language-jlcon">julia&gt; record = VCF.Record(&quot;20\t14370\trs6054257\tG\tA\t29\tPASS\tNS=3;DP=14;AF=0.5;DB;H2\tGT:GQ:DP:HQ\t0|0:48:1:51,51\t1|0:48:8:51,51&quot;)
GeneticVariation.VCF.Record:
   chromosome: 20
     position: 14370
   identifier: rs6054257
    reference: G
    alternate: A
      quality: 29.0
       filter: PASS
  information: NS=3 DP=14 AF=0.5 DB H2
       format: GT GQ DP HQ
     genotype: [1] 0|0 48 1 51,51 [2] 1|0 48 8 51,51

julia&gt; VCF.chrom(record)
&quot;20&quot;

julia&gt; VCF.pos(record)
14370

julia&gt; VCF.id(record)
1-element Array{String,1}:
 &quot;rs6054257&quot;

julia&gt; VCF.ref(record)
&quot;G&quot;

julia&gt; VCF.alt(record)
1-element Array{String,1}:
 &quot;A&quot;

julia&gt; VCF.qual(record)
29.0

julia&gt; VCF.filter(record)
1-element Array{String,1}:
 &quot;PASS&quot;

julia&gt; VCF.info(record)
5-element Array{Pair{String,String},1}:
 &quot;NS&quot;=&gt;&quot;3&quot;  
 &quot;DP&quot;=&gt;&quot;14&quot;
 &quot;AF&quot;=&gt;&quot;0.5&quot;
 &quot;DB&quot;=&gt;&quot;&quot;   
 &quot;H2&quot;=&gt;&quot;&quot;   

julia&gt; VCF.format(record)
4-element Array{String,1}:
 &quot;GT&quot;
 &quot;GQ&quot;
 &quot;DP&quot;
 &quot;HQ&quot;

julia&gt; VCF.genotype(record)
2-element Array{Array{String,1},1}:
 String[&quot;0|0&quot;,&quot;48&quot;,&quot;1&quot;,&quot;51,51&quot;]
 String[&quot;1|0&quot;,&quot;48&quot;,&quot;8&quot;,&quot;51,51&quot;]

julia&gt; VCF.genotype(record, 1:2, &quot;GT&quot;)
2-element Array{String,1}:
 &quot;0|0&quot;
 &quot;1|0&quot;

julia&gt; VCF.genotype(record, 1:1, &quot;GT&quot;)
1-element Array{String,1}:
 &quot;0|0&quot;

julia&gt; VCF.genotype(record, 1:2, &quot;GT&quot;)
2-element Array{String,1}:
 &quot;0|0&quot;
 &quot;1|0&quot;
</code></pre><footer><hr/><a class="previous" href="../../../"><span class="direction">Previous</span><span class="title">Home</span></a><a class="next" href="../../site_counting/"><span class="direction">Next</span><span class="title">Mutation Counting</span></a></footer></article></body></html>
