<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Docstrings · Kpax3.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>Kpax3.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../">Readme</a></li><li class="current"><a class="toctext" href>Docstrings</a><ul class="internal"></ul></li></ul></nav><article id="docs"><header><nav><ul><li><a href>Docstrings</a></li></ul></nav><hr/><div id="topbar"><span>Docstrings</span><a class="fa fa-bars" href="#"></a></div></header><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.AminoAcidData" href="#Kpax3.AminoAcidData"><code>Kpax3.AminoAcidData</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p><strong>Genetic data</strong></p><p><strong>Description</strong></p><p>Amino acid data and its metadata.</p><p><strong>Fields</strong></p><ul><li><code>data</code> Multiple sequence alignment (MSA) encoded as a binary (UInt8) matrix</li><li><code>id</code> units&#39; ids</li><li><code>ref</code> reference sequence, i.e. a vector of the same length of the original</li></ul><p>sequences storing the values of homogeneous sites. SNPs are instead represented by a value of 29</p><ul><li><code>val</code> vector with unique values per MSA site</li><li><code>key</code> vector with indices of each value</li></ul><p><strong>Details</strong></p><p>Let <code>n</code> be the total number of units and <code>ml</code> be the total number of unique values observed at SNP <code>l</code>. Define m = m1 + ... + mL, where L is the total number of SNPs.</p><p><code>data</code> is a <code>m</code>-by-<code>n</code> indicator matrix, i.e. <code>data[j, i]</code> is <code>1</code> if unit <code>i</code> possesses value <code>j</code>, <code>0</code> otherwise.</p><p>The value associated with column <code>j</code> can be obtained by <code>val[j]</code> while the SNP position by <code>findall(ref == 29)[key[j]]</code>.</p><p><strong>References</strong></p><p>Pessia A., Grad Y., Cobey S., Puranen J. S. and Corander J. (2015). K-Pax2: Bayesian identification of cluster-defining amino acid positions in large sequence datasets. <em>Microbial Genomics</em> <strong>1</strong>(1). <a href="http://dx.doi.org/10.1099/mgen.0.000025">http://dx.doi.org/10.1099/mgen.0.000025</a>.</p></div></div></section><pre><code class="language-none">Kpax3.AminoAcidPriorCol</code></pre><pre><code class="language-none">Kpax3.AminoAcidState</code></pre><pre><code class="language-none">Kpax3.AminoAcidStateList</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.CategoricalData" href="#Kpax3.CategoricalData"><code>Kpax3.CategoricalData</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p><strong>CSV data</strong></p><p><strong>Description</strong></p><p>Generic data and its metadata.</p><p><strong>Fields</strong></p><ul><li><code>data</code> original data matrix encoded as a binary (UInt8) matrix</li><li><code>id</code> units&#39; ids</li><li><code>ref</code> reference observation, i.e. a vector of the same length of the original</li></ul><p>observations storing the values of homogeneous sites. Polymorphisms are instead represented by the string &quot;.&quot;</p><ul><li><code>val</code> vector with unique values per dataset column</li><li><code>key</code> vector with indices of each value</li></ul><p><strong>Details</strong></p><p>Let <code>n</code> be the total number of units and <code>ml</code> be the total number of unique values observed at polymorphic column <code>l</code>. Define m = m1 + ... + mL, where L is the total number of polymorphic columns.</p><p><code>data</code> is a <code>m</code>-by-<code>n</code> indicator matrix, i.e. <code>data[j, i]</code> is <code>1</code> if unit <code>i</code> possesses value <code>j</code>, <code>0</code> otherwise.</p><p>The value associated with column <code>j</code> can be obtained by <code>val[j]</code> while the polymorphic position by <code>findall(ref == &quot;.&quot;)[key[j]]</code>.</p><p><strong>References</strong></p><p>Pessia A., Grad Y., Cobey S., Puranen J. S. and Corander J. (2015). K-Pax2: Bayesian identification of cluster-defining amino acid positions in large sequence datasets. <em>Microbial Genomics</em> <strong>1</strong>(1). <a href="http://dx.doi.org/10.1099/mgen.0.000025">http://dx.doi.org/10.1099/mgen.0.000025</a>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.EwensPitman" href="#Kpax3.EwensPitman"><code>Kpax3.EwensPitman</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p><strong>Ewens-Pitman distribution</strong></p><p><strong>Description</strong></p><p>The Ewens-Pitman distribution is a discrete probability distribution on the partitions of the set N = {1, ..., n} (n = 1, 2, ...). It is a two parameters generalization of the Ewens sampling formula. The latter is also known as the Chinese Restaurant Process or as the marginal distribution of a Dirichlet Process.</p><p><strong>Fields</strong></p><ul><li><code>α</code> Real number (see details)</li><li><code>θ</code> Real number (see details)</li></ul><p><strong>Details</strong></p><p>The two parameters must satisfy either</p><ul><li>α &lt; 0 and θ = -Lα for some L ∈ {1, 2, ...}</li><li>0 ≤ α &lt; 1 and θ &gt; − α</li></ul><p>The special case α = 0 and θ &gt; 0 corresponds to the Ewens sampling formula.</p><p><strong>References</strong></p><p>Aldous, D. J. (1985) Exchangeability and related topics. In <em>École d&#39;Été de Probabilités de Saint-Flour XIII — 1983</em>. Lecture Notes in Mathematics <strong>1117</strong>, 1-198. Springer Berlin Heidelberg. <a href="http://dx.doi.org/10.1007/BFb0099421">http://dx.doi.org/10.1007/BFb0099421</a>.</p><p>Gnedin, A. and Pitman, J. (2006) Exchangeable Gibbs partitions and Stirling triangles. <em>Journal of Mathematical Sciences</em> <strong>138</strong>(3), 5674-5685. <a href="http://dx.doi.org/10.1007/s10958-006-0335-z">http://dx.doi.org/10.1007/s10958-006-0335-z</a>.</p><p>Kerov, S. (2006) Coherent random allocations, and the Ewens-Pitman formula. <em>Journal of Mathematical Sciences</em>, <strong>138</strong>(3). <a href="http://dx.doi.org/10.1007/s10958-006-0338-9">http://dx.doi.org/10.1007/s10958-006-0338-9</a>.</p><p>Pitman, J. (1995) Exchangeable and Partially Exchangeable Random Partitions. <em>Probability Theory and Related Fields</em> <strong>102</strong>(2), 145-158. <a href="http://dx.doi.org/10.1007%2FBF01213386">http://dx.doi.org/10.1007%2FBF01213386</a>.</p><p>Pitman, J. (2006) Combinatorial Stochastic Processes. In <em>Ecole d’Eté de Probabilités de Saint-Flour XXXII – 2002</em>. Lecture Notes in Mathematics <strong>1875</strong>. Springer Berlin Heidelberg. <a href="http://dx.doi.org/10.1007/b11601500">http://dx.doi.org/10.1007/b11601500</a>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.EwensPitmanNAPT" href="#Kpax3.EwensPitmanNAPT"><code>Kpax3.EwensPitmanNAPT</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p><strong>Ewens-Pitman distribution</strong></p><p><strong>Description</strong></p><p>Ewens-Pitman distribution with α &lt; 0 and θ &gt; 0. θ = -Lα for some L ∈ {1, 2, ...}.</p><p><strong>Fields</strong></p><ul><li><code>α</code> Real number lesser than zero</li><li><code>L</code> Integer number greater than zero</li></ul></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.EwensPitmanPAUT" href="#Kpax3.EwensPitmanPAUT"><code>Kpax3.EwensPitmanPAUT</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p><strong>Ewens-Pitman distribution</strong></p><p><strong>Description</strong></p><p>Ewens-Pitman distribution with 0 &lt; α &lt; 1,  θ &gt; -α and θ ≠ 0.</p><p><strong>Fields</strong></p><ul><li><code>α</code> Real number greater than zero and lesser than one</li><li><code>θ</code> Real number greater than <code>-α</code> but different from zero</li></ul></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.EwensPitmanPAZT" href="#Kpax3.EwensPitmanPAZT"><code>Kpax3.EwensPitmanPAZT</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p><strong>Ewens-Pitman distribution</strong></p><p><strong>Description</strong></p><p>Ewens-Pitman distribution with 0 &lt; α &lt; 1 and θ = 0.</p><p><strong>Fields</strong></p><ul><li><code>α</code> Real number greater than zero and lesser than one</li></ul></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.EwensPitmanZAPT" href="#Kpax3.EwensPitmanZAPT"><code>Kpax3.EwensPitmanZAPT</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p><strong>Ewens-Pitman distribution</strong></p><p><strong>Description</strong></p><p>Ewens-Pitman distribution with α = 0 and θ &gt; 0. This is equivalent to the Ewens sampling formula.</p><p><strong>Fields</strong></p><ul><li><code>θ</code> Real number greater than zero</li></ul></div></div></section><pre><code class="language-none">Kpax3.GASupport</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.KCSVError" href="#Kpax3.KCSVError"><code>Kpax3.KCSVError</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p><strong>Kpax3 Exception</strong></p><p><strong>Description</strong></p><p>Exception for a wrong formatted CSV file.</p><p><strong>Fields</strong></p><ul><li><code>msg</code> Optional argument with a descriptive error string</li></ul></div></div></section><pre><code class="language-none">Kpax3.KData</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.KDomainError" href="#Kpax3.KDomainError"><code>Kpax3.KDomainError</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p><strong>Kpax3 Exception</strong></p><p><strong>Description</strong></p><p>Provides a message explaining the reason of the DomainError exception.</p><p><strong>Fields</strong></p><ul><li><code>msg</code> Optional argument with a descriptive error string</li></ul></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.KFASTAError" href="#Kpax3.KFASTAError"><code>Kpax3.KFASTAError</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p><strong>Kpax3 Exception</strong></p><p><strong>Description</strong></p><p>Exception for a wrong formatted FASTA file.</p><p><strong>Fields</strong></p><ul><li><code>msg</code> Optional argument with a descriptive error string</li></ul></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.KInputError" href="#Kpax3.KInputError"><code>Kpax3.KInputError</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p><strong>Kpax3 Exception</strong></p><p><strong>Description</strong></p><p>Exception for wrong data read from a source.</p><p><strong>Fields</strong></p><ul><li><code>msg</code> Optional argument with a descriptive error string</li></ul></div></div></section><pre><code class="language-none">Kpax3.KOffspring</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.KSettings" href="#Kpax3.KSettings"><code>Kpax3.KSettings</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p><strong>User defined settings for a Kpax3 run</strong></p><p><strong>Description</strong></p><p><strong>Fields</strong></p></div></div></section><pre><code class="language-none">Kpax3.KWeight</code></pre><pre><code class="language-none">Kpax3.Kpax3</code></pre><pre><code class="language-none">Kpax3.Kpax3PlotC</code></pre><pre><code class="language-none">Kpax3.Kpax3PlotD</code></pre><pre><code class="language-none">Kpax3.Kpax3PlotDensity</code></pre><pre><code class="language-none">Kpax3.Kpax3PlotJump</code></pre><pre><code class="language-none">Kpax3.Kpax3PlotK</code></pre><pre><code class="language-none">Kpax3.Kpax3PlotP</code></pre><pre><code class="language-none">Kpax3.Kpax3PlotTrace</code></pre><pre><code class="language-none">Kpax3.MCMCSupport</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.NucleotideData" href="#Kpax3.NucleotideData"><code>Kpax3.NucleotideData</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p><strong>Genetic data</strong></p><p><strong>Description</strong></p><p>DNA data and its metadata.</p><p><strong>Fields</strong></p><ul><li><code>data</code> Multiple sequence alignment (MSA) encoded as a binary (UInt8) matrix</li><li><code>id</code> units&#39; ids</li><li><code>ref</code> reference sequence, i.e. a vector of the same length of the original</li></ul><p>sequences storing the values of homogeneous sites. SNPs are instead represented by a value of 29</p><ul><li><code>val</code> vector with unique values per MSA site</li><li><code>key</code> vector with indices of each value</li></ul><p><strong>Details</strong></p><p>Let <code>n</code> be the total number of units and <code>ml</code> be the total number of unique values observed at SNP <code>l</code>. Define m = m1 + ... + mL, where L is the total number of SNPs.</p><p><code>data</code> is a <code>m</code>-by-<code>n</code> indicator matrix, i.e. <code>data[j, i]</code> is <code>1</code> if unit <code>i</code> possesses value <code>j</code>, <code>0</code> otherwise.</p><p>The value associated with column <code>j</code> can be obtained by <code>val[j]</code> while the SNP position by <code>findall(ref == 29)[key[j]]</code>.</p><p><strong>References</strong></p><p>Pessia A., Grad Y., Cobey S., Puranen J. S. and Corander J. (2015). K-Pax2: Bayesian identification of cluster-defining amino acid positions in large sequence datasets. <em>Microbial Genomics</em> <strong>1</strong>(1). <a href="http://dx.doi.org/10.1099/mgen.0.000025">http://dx.doi.org/10.1099/mgen.0.000025</a>.</p></div></div></section><pre><code class="language-none">Kpax3.PriorColPartition</code></pre><pre><code class="language-none">Kpax3.PriorRowPartition</code></pre><pre><code class="language-none">Kpax3.State</code></pre><pre><code class="language-none">Kpax3.StateList</code></pre><pre><code class="language-none">Kpax3.aamtn84</code></pre><pre><code class="language-none">Kpax3.biased_random_walk!</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.categorical2binary" href="#Kpax3.categorical2binary"><code>Kpax3.categorical2binary</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p><strong>Convert categorical (integer) data to binary</strong></p><p><strong>Description</strong></p><p>Convert an integer matrix to a binary (indicator) matrix.</p><p><strong>Usage</strong></p><p>categorical2binary(data, maxval, missval)</p><p><strong>Arguments</strong></p><ul><li><code>data</code> Integer matrix to be converted</li><li><code>maxval</code> Theoretical maximum value observable in <code>data</code></li><li><code>missval</code> Value to be considered missing</li></ul><p><strong>Value</strong></p><p>A tuple containing the following variables:</p><ul><li><code>bindata</code> Original data matrix encoded as a binary (indicator) matrix</li><li><code>val</code> vector with unique values per MSA site</li><li><code>key</code> vector with indices of each value</li></ul><p><strong>Example</strong></p><p>If <code>data</code> consists of just the following three units</p><pre><code class="language-none">0 2 1
1 3 2
2 4 0
2 2 3</code></pre><p>then <code>bindata</code> will be equal to</p><pre><code class="language-none">0 0 1
0 1 0
1 0 0
0 0 1
0 1 0
1 0 0
0 1 0
1 1 0
0 0 1</code></pre><p>while</p><pre><code class="language-none">val = [1, 2, 1, 2, 3, 2, 4, 2, 3] (i.e. 12 123 24 23)
key = [1, 1, 2, 2, 2, 3, 3, 4, 4] (i.e. 11 222 33 44)</code></pre><p><code>0</code> values (here missing data) are discarded.</p></div></div><div><div><p><strong>Convert categorical (string) data to binary</strong></p><p><strong>Description</strong></p><p>Convert a string matrix to a binary (indicator) matrix.</p><p><strong>Usage</strong></p><p>categorical2binary(data, missval)</p><p><strong>Arguments</strong></p><ul><li><code>data</code> Integer matrix to be converted</li><li><code>missval</code> Value to be considered missing</li></ul><p><strong>Value</strong></p><p>A tuple containing the following variables:</p><ul><li><code>bindata</code> Original data matrix encoded as a binary (indicator) matrix</li><li><code>val</code> vector with unique values per MSA site</li><li><code>key</code> vector with indices of each value</li></ul><p><strong>Example</strong></p><p>If <code>data</code> consists of just the following three units</p><pre><code class="language-none">  C A
A G C
C T
C C G</code></pre><p>then <code>bindata</code> will be equal to</p><pre><code class="language-none">0 0 1
0 1 0
1 0 0
0 0 1
0 1 0
1 0 0
0 1 0
1 1 0
0 0 1</code></pre><p>while</p><pre><code class="language-none">val = [&quot;A&quot;, &quot;C&quot;, &quot;A&quot;, &quot;C&quot;, &quot;G, &quot;C&quot;, &quot;T&quot;, &quot;C&quot;, &quot;G&quot;] (i.e. AC ACG CT CG)
key = [ 1,   1,   2,   2,   2,  3,   3,   4,   4 ] (i.e. 11 222 33 44)</code></pre><p><code></code> values (here missing data) are discarded.</p></div></div></section><pre><code class="language-none">Kpax3.clusterweight</code></pre><pre><code class="language-none">Kpax3.computeclusteriseqprobs!</code></pre><pre><code class="language-none">Kpax3.computeclusterjseqprobs!</code></pre><pre><code class="language-none">Kpax3.computeclusterlogprobs!</code></pre><pre><code class="language-none">Kpax3.computelocalmode!</code></pre><pre><code class="language-none">Kpax3.condmarglik</code></pre><pre><code class="language-none">Kpax3.copystate</code></pre><pre><code class="language-none">Kpax3.copystate!</code></pre><pre><code class="language-none">Kpax3.copystatelist!</code></pre><pre><code class="language-none">Kpax3.countfields</code></pre><pre><code class="language-none">Kpax3.crossover!</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.dPriorRow" href="#Kpax3.dPriorRow"><code>Kpax3.dPriorRow</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p><strong>Density of the Ewens-Pitman distribution</strong></p><p><strong>Description</strong></p><p>Probability of a partition according to the Ewens-Pitman distribution.</p><p><strong>Usage</strong></p><p>dPriorRow(ep, p) dPriorRow(ep, n, k, m)</p><p><strong>Arguments</strong></p><ul><li><code>ep</code> Object of (super)type EwensPitman</li><li><code>p</code> Vector of integers representing a partition</li><li><code>n</code> Set size (Integer)</li><li><code>k</code> Number of blocks (Integer)</li><li><code>m</code> Vector of integers representing block sizes</li></ul><p><strong>Details</strong></p><p><strong>Examples</strong></p></div></div></section><pre><code class="language-none">Kpax3.decodepartition!</code></pre><pre><code class="language-none">Kpax3.distaamtn84</code></pre><pre><code class="language-none">Kpax3.distntmtn84</code></pre><pre><code class="language-none">Kpax3.distntmtn93</code></pre><pre><code class="language-none">Kpax3.distnttn84</code></pre><pre><code class="language-none">Kpax3.distnttn93</code></pre><pre><code class="language-none">Kpax3.distsj</code></pre><pre><code class="language-none">Kpax3.distsjlower</code></pre><pre><code class="language-none">Kpax3.distsjupper</code></pre><pre><code class="language-none">Kpax3.encodepartition!</code></pre><pre><code class="language-none">Kpax3.ess</code></pre><pre><code class="language-none">Kpax3.eval</code></pre><pre><code class="language-none">Kpax3.expandrect</code></pre><pre><code class="language-none">Kpax3.expandrecthoriz</code></pre><pre><code class="language-none">Kpax3.expandrectverti</code></pre><pre><code class="language-none">Kpax3.expandsquare</code></pre><pre><code class="language-none">Kpax3.expandsquarediag</code></pre><pre><code class="language-none">Kpax3.gibbs!</code></pre><pre><code class="language-none">Kpax3.gibbsallocate!</code></pre><pre><code class="language-none">Kpax3.gibbscomputenormconst</code></pre><pre><code class="language-none">Kpax3.gibbscomputeprobg!</code></pre><pre><code class="language-none">Kpax3.gibbscomputeprobi!</code></pre><pre><code class="language-none">Kpax3.gibbscomputeprobj!</code></pre><pre><code class="language-none">Kpax3.gibbsmerge!</code></pre><pre><code class="language-none">Kpax3.gibbsmove!</code></pre><pre><code class="language-none">Kpax3.gibbssamplecluster</code></pre><pre><code class="language-none">Kpax3.gibbssplit!</code></pre><pre><code class="language-none">Kpax3.gibbsupdateclustergmerge!</code></pre><pre><code class="language-none">Kpax3.gibbsupdateclustergmove!</code></pre><pre><code class="language-none">Kpax3.gibbsupdateclusteri!</code></pre><pre><code class="language-none">Kpax3.gibbsupdateclusterj!</code></pre><pre><code class="language-none">Kpax3.imsevar</code></pre><pre><code class="language-none">Kpax3.include</code></pre><pre><code class="language-none">Kpax3.initclusteriweights!</code></pre><pre><code class="language-none">Kpax3.initclusterjweights!</code></pre><pre><code class="language-none">Kpax3.initializepartition</code></pre><pre><code class="language-none">Kpax3.initializestate</code></pre><pre><code class="language-none">Kpax3.initsupportbrwmerge!</code></pre><pre><code class="language-none">Kpax3.initsupportbrwmove!</code></pre><pre><code class="language-none">Kpax3.initsupportbrwsplit!</code></pre><pre><code class="language-none">Kpax3.initsupportgibbs!</code></pre><pre><code class="language-none">Kpax3.initsupportmerge!</code></pre><pre><code class="language-none">Kpax3.initsupportsplit!</code></pre><pre><code class="language-none">Kpax3.jaccard</code></pre><pre><code class="language-none">Kpax3.jaccard_index_basic</code></pre><pre><code class="language-none">Kpax3.jaccard_index_table</code></pre><pre><code class="language-none">Kpax3.jaccardlower</code></pre><pre><code class="language-none">Kpax3.kpax3Restimate</code></pre><pre><code class="language-none">Kpax3.kpax3estimate</code></pre><pre><code class="language-none">Kpax3.kpax3ga</code></pre><pre><code class="language-none">Kpax3.kpax3ga!</code></pre><pre><code class="language-none">Kpax3.kpax3mcmc</code></pre><pre><code class="language-none">Kpax3.kpax3mcmc!</code></pre><pre><code class="language-none">Kpax3.kpax3plotc</code></pre><pre><code class="language-none">Kpax3.kpax3plotc!</code></pre><pre><code class="language-none">Kpax3.kpax3plotd</code></pre><pre><code class="language-none">Kpax3.kpax3plotd!</code></pre><pre><code class="language-none">Kpax3.kpax3plotdensity</code></pre><pre><code class="language-none">Kpax3.kpax3plotdensity!</code></pre><pre><code class="language-none">Kpax3.kpax3plotjump</code></pre><pre><code class="language-none">Kpax3.kpax3plotjump!</code></pre><pre><code class="language-none">Kpax3.kpax3plotk</code></pre><pre><code class="language-none">Kpax3.kpax3plotk!</code></pre><pre><code class="language-none">Kpax3.kpax3plotp</code></pre><pre><code class="language-none">Kpax3.kpax3plotp!</code></pre><pre><code class="language-none">Kpax3.kpax3plottrace</code></pre><pre><code class="language-none">Kpax3.kpax3plottrace!</code></pre><pre><code class="language-none">Kpax3.loadaa</code></pre><pre><code class="language-none">Kpax3.loadnt</code></pre><pre><code class="language-none">Kpax3.logcondmarglik</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.logcondpostC" href="#Kpax3.logcondpostC"><code>Kpax3.logcondpostC</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Logp(C | R, X)</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.logcondpostS" href="#Kpax3.logcondpostS"><code>Kpax3.logcondpostS</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>logp(S | R, X)</p></div></div></section><pre><code class="language-none">Kpax3.logdPriorRow</code></pre><pre><code class="language-none">Kpax3.loglikelihood</code></pre><pre><code class="language-none">Kpax3.logmarglik</code></pre><pre><code class="language-none">Kpax3.logmarglikbrwmerge!</code></pre><pre><code class="language-none">Kpax3.logmarglikbrwmove!</code></pre><pre><code class="language-none">Kpax3.logmarglikbrwsplit!</code></pre><pre><code class="language-none">Kpax3.logmarglikelihood</code></pre><pre><code class="language-none">Kpax3.logmarglikmerge!</code></pre><pre><code class="language-none">Kpax3.logmargliksplit!</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.logposterior" href="#Kpax3.logposterior"><code>Kpax3.logposterior</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Logposterior minus the log(normalizing constant)</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.logpriorC" href="#Kpax3.logpriorC"><code>Kpax3.logpriorC</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>logp(C | R)</p></div></div></section><pre><code class="language-none">Kpax3.logratiopriorrowmerge</code></pre><pre><code class="language-none">Kpax3.logratiopriorrowmerge!</code></pre><pre><code class="language-none">Kpax3.logratiopriorrowmove</code></pre><pre><code class="language-none">Kpax3.logratiopriorrowsplit</code></pre><pre><code class="language-none">Kpax3.logratiopriorrowsplit!</code></pre><pre><code class="language-none">Kpax3.loss_binder</code></pre><pre><code class="language-none">Kpax3.marglik</code></pre><pre><code class="language-none">Kpax3.merge!</code></pre><pre><code class="language-none">Kpax3.modifymerge!</code></pre><pre><code class="language-none">Kpax3.modifypartition!</code></pre><pre><code class="language-none">Kpax3.modifyscramble!</code></pre><pre><code class="language-none">Kpax3.modifysplit!</code></pre><pre><code class="language-none">Kpax3.mutation!</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.normalizepartition" href="#Kpax3.normalizepartition"><code>Kpax3.normalizepartition</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>remove &quot;gaps&quot; and non-positive values from the partition Example: [1, 1, 0, 1, -2, 4, 0] -&gt; [3, 3, 2, 3, 1, 4, 2]</p></div></div></section><pre><code class="language-none">Kpax3.ntmtn84</code></pre><pre><code class="language-none">Kpax3.ntmtn93</code></pre><pre><code class="language-none">Kpax3.nttn84</code></pre><pre><code class="language-none">Kpax3.nttn93</code></pre><pre><code class="language-none">Kpax3.optimumstate</code></pre><pre><code class="language-none">Kpax3.performbrwmerge!</code></pre><pre><code class="language-none">Kpax3.performbrwmove!</code></pre><pre><code class="language-none">Kpax3.performbrwsplit!</code></pre><pre><code class="language-none">Kpax3.performmerge!</code></pre><pre><code class="language-none">Kpax3.performsplit!</code></pre><pre><code class="language-none">Kpax3.plotC</code></pre><pre><code class="language-none">Kpax3.plotD</code></pre><pre><code class="language-none">Kpax3.plotP</code></pre><pre><code class="language-none">Kpax3.plotdensity</code></pre><pre><code class="language-none">Kpax3.plotdgn</code></pre><pre><code class="language-none">Kpax3.plotjump</code></pre><pre><code class="language-none">Kpax3.plotk</code></pre><pre><code class="language-none">Kpax3.plottrace</code></pre><pre><code class="language-none">Kpax3.processchain</code></pre><pre><code class="language-none">Kpax3.readdata</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="Kpax3.readfasta" href="#Kpax3.readfasta"><code>Kpax3.readfasta</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>readfasta(ifile::String, protein::Bool, miss::Vector{UInt8}, l::Int, verbose::Bool, verbosestep::Int)</p><p>Read data in FASTA format and convert it to an integer matrix. Sequences are required to be aligned. Only polymorphic columns are stored.</p><p><strong>Arguments</strong></p><ul><li><code>ifile</code> Path to the input data file</li><li><code>protein</code> <code>true</code> if reading protein data or <code>false</code> if reading DNA data</li><li><code>miss</code> Characters (as <code>UInt8</code>) to be considered missing. Use</li></ul><p><code>miss = zeros(UInt8, 1)</code> if all characters are to be considered valid. Default characters for <code>miss</code> are:</p><pre><code class="language-none">- DNA data: _?, \*, #, -, b, d, h, k, m, n, r, s, v, w, x, y, j, z_
- Protein data: _?, \*, #, -, b, j, x, z_</code></pre><ul><li><code>l</code> Sequence length. If unknown, it is better to choose a value which is</li></ul><p>surely greater than the real sequence length. If <code>l</code> is found to be insufficient, the array size is dynamically increased (not recommended from a performance point of view). Default value should be sufficient for most datasets</p><ul><li><code>verbose</code> If <code>true</code>, print status reports</li><li><code>verbosestep</code> Print a status report every <code>verbosestep</code> read sequences</li></ul><p><strong>Details</strong></p><p>When computing evolutionary distances, don&#39;t put the gap symbol <code>-</code> among the missing values. Indeed, indels are an important piece of information for genetic distances.</p><p>FASTA data is encoded as standard 7-bit ASCII codes. The only exception is Uracil which is given the same value 84 of Thymidine, i.e. each &#39;u&#39; is silently converted to &#39;t&#39; when reading DNA data. Conversion tables are the following:</p><p>+––––––––––––––––––––+ |         Conversion table (DNA)         | +––––––––––––––––––––+ |          Nucleotide |  Code |  Integer | +––––––––––-+–––-+–––––+ |           Adenosine |   A   |    97    | |            Cytosine |   C   |    99    | |             Guanine |   G   |   103    | |           Thymidine |   T   |   116    | |              Uracil |   U   |   116    | |     Purine (A or G) |   R   |   114    | | Pyrimidine (C or T) |   Y   |   121    | |                Keto |   K   |   107    | |               Amino |   M   |   109    | |  Strong Interaction |   S   |   115    | |    Weak Interaction |   W   |   119    | |               Not A |   B   |    98    | |               Not C |   D   |   100    | |               Not G |   H   |   104    | |          Not T or U |   V   |   118    | |                 Any |   N   |   110    | |                 Gap |   -   |    45    | |              Masked |   X   |   120    | +––––––––––––––––––––+</p><p>+––––––––––––––––––––––––+ |           Conversion table (PROTEIN)           | +––––––––––––––––––––––––+ |                  Amino Acid |  Code |  Integer | +––––––––––––––-+–––-+–––––+ |                     Alanine |   A   |    97    | |                    Arginine |   R   |   114    | |                  Asparagine |   N   |   110    | |               Aspartic acid |   D   |   100    | |                    Cysteine |   C   |    99    | |                   Glutamine |   Q   |   113    | |               Glutamic acid |   E   |   101    | |                     Glycine |   G   |   103    | |                   Histidine |   H   |   104    | |                  Isoleucine |   I   |   105    | |                     Leucine |   L   |   108    | |                      Lysine |   K   |   107    | |                  Methionine |   M   |   109    | |               Phenylalanine |   F   |   102    | |                     Proline |   P   |   112    | |                 Pyrrolysine |   O   |   111    | |              Selenocysteine |   U   |   117    | |                      Serine |   S   |   115    | |                   Threonine |   T   |   116    | |                  Tryptophan |   W   |   119    | |                    Tyrosine |   Y   |   121    | |                      Valine |   V   |   118    | | Asparagine or Aspartic acid |   B   |    98    | |  Glutamine or Glutamic acid |   Z   |   122    | |       Leucine or Isoleucine |   J   |   106    | |                         Gap |   -   |    45    | |            Translation stop |   *   |    42    | |                         Any |   X   |   120    | +––––––––––––––––––––––––+</p><p><strong>Value</strong></p><p>A tuple containing the following variables:</p><ul><li><code>data</code> Multiple Sequence Alignment (MSA) encoded as a UInt8 matrix</li><li><code>id</code> Units&#39; ids</li><li><code>ref</code> Reference sequence, i.e. a vector of the same length of the original</li></ul><p>sequences storing the values of homogeneous sites. SNPs are instead represented by a value of 46 (&#39;.&#39;)</p></div></div></section><pre><code class="language-none">Kpax3.readposteriorC</code></pre><pre><code class="language-none">Kpax3.readposteriorP</code></pre><pre><code class="language-none">Kpax3.readposteriork</code></pre><pre><code class="language-none">Kpax3.reorderunits</code></pre><pre><code class="language-none">Kpax3.resizestate!</code></pre><pre><code class="language-none">Kpax3.resizesupport!</code></pre><pre><code class="language-none">Kpax3.rpostpartitioncols!</code></pre><pre><code class="language-none">Kpax3.sampleC!</code></pre><pre><code class="language-none">Kpax3.samplenoise!</code></pre><pre><code class="language-none">Kpax3.samplestrongsignal!</code></pre><pre><code class="language-none">Kpax3.sampleweaksignal!</code></pre><pre><code class="language-none">Kpax3.save</code></pre><pre><code class="language-none">Kpax3.savestate!</code></pre><pre><code class="language-none">Kpax3.selection</code></pre><pre><code class="language-none">Kpax3.shuffleunits!</code></pre><pre><code class="language-none">Kpax3.split!</code></pre><pre><code class="language-none">Kpax3.splitmerge!</code></pre><pre><code class="language-none">Kpax3.traceC</code></pre><pre><code class="language-none">Kpax3.traceR</code></pre><pre><code class="language-none">Kpax3.updateclusteri!</code></pre><pre><code class="language-none">Kpax3.updateclusteriweights!</code></pre><pre><code class="language-none">Kpax3.updateclusterj!</code></pre><pre><code class="language-none">Kpax3.updateclusterjweights!</code></pre><pre><code class="language-none">Kpax3.updatelogmargliki!</code></pre><pre><code class="language-none">Kpax3.updatelogmarglikj!</code></pre><pre><code class="language-none">Kpax3.updatestate!</code></pre><pre><code class="language-none">Kpax3.writeattributes</code></pre><pre><code class="language-none">Kpax3.writecvalues</code></pre><pre><code class="language-none">Kpax3.writedataset</code></pre><pre><code class="language-none">Kpax3.writeresults</code></pre><footer><hr/><a class="previous" href="../"><span class="direction">Previous</span><span class="title">Readme</span></a></footer></article></body></html>
