<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>FASTQ formatted files · BioSequences.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="../.."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../../assets/documenter.js"></script><script src="../../siteinfo.js"></script><script src="../../../versions.js"></script><link href="../../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><a href="../../index.html"><img class="logo" src="../../assets/logo.png" alt="BioSequences.jl logo"/></a><h1>BioSequences.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../../">Home</a></li><li><a class="toctext" href="../../symbols/">Biological Symbols</a></li><li><span class="toctext">Sequence types</span><ul><li><a class="toctext" href="../../sequences/sequences/">Overview</a></li><li><a class="toctext" href="../../sequences/bioseq/">BioSequence</a></li><li><a class="toctext" href="../../sequences/refseq/">Reference Sequences</a></li><li><a class="toctext" href="../../sequences/kmer/">Nucleic acid k-mers</a></li></ul></li><li><span class="toctext">IO</span><ul><li><a class="toctext" href="../fasta/">FASTA formatted files</a></li><li class="current"><a class="toctext" href>FASTQ formatted files</a><ul class="internal"><li><a class="toctext" href="#Readers-and-Writers-1">Readers and Writers</a></li></ul></li><li><a class="toctext" href="../twobit/">2bit formatted files</a></li></ul></li><li><a class="toctext" href="../../search/">Searching</a></li><li><a class="toctext" href="../../composition/">Sequence Composition</a></li><li><a class="toctext" href="../../demultiplexer/">Demultiplexing</a></li><li><a class="toctext" href="../../contributing/">Contributing</a></li></ul></nav><article id="docs"><header><nav><ul><li>IO</li><li><a href>FASTQ formatted files</a></li></ul></nav><hr/><div id="topbar"><span>FASTQ formatted files</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="IO-FASTQ-formatted-files-1" href="#IO-FASTQ-formatted-files-1">IO - FASTQ formatted files</a></h1><p>FASTQ is a text-based file format for representing DNA sequences along with qualities for each base. A FASTQ file stores a list of sequence records in the following format:</p><pre><code class="language-none">@{name} {description}?
{sequence}
+
{qualities}</code></pre><p>Here is an example of one record from a FASTQ file:</p><pre><code class="language-none">@FSRRS4401BE7HA
tcagTTAAGATGGGAT
+
###EEEEEEEEE##E#</code></pre><h2><a class="nav-anchor" id="Readers-and-Writers-1" href="#Readers-and-Writers-1">Readers and Writers</a></h2><p>The reader and writer for FASTQ formatted files, are found within the <code>BioSequences.FASTQ</code> module.</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.FASTQ.Reader" href="#BioSequences.FASTQ.Reader"><code>BioSequences.FASTQ.Reader</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">FASTQ.Reader(input::IO; fill_ambiguous=nothing)</code></pre><p>Create a data reader of the FASTQ file format.</p><p><strong>Arguments</strong></p><ul><li><code>input</code>: data source</li><li><code>fill_ambiguous=nothing</code>: fill ambiguous symbols with the given symbol</li></ul></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.FASTQ.Writer" href="#BioSequences.FASTQ.Writer"><code>BioSequences.FASTQ.Writer</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">FASTQ.Writer(output::IO; quality_header=false)</code></pre><p>Create a data writer of the FASTQ file format.</p><p><strong>Arguments</strong></p><ul><li><code>output</code>: data sink</li><li><code>quality_header=false</code>: output the title line at the third line just after &#39;+&#39;</li></ul></div></div></section><p>They can be created with IOStreams:</p><pre><code class="language-jlcon">r = FASTQ.Reader(open(&quot;MyInput.fastq&quot;, &quot;r&quot;))
w = FASTQ.Writer(open(&quot;MyFile.fastq&quot;, &quot;w&quot;))</code></pre><p>Note that <code>FASTQ.Reader</code> does not support line-wraps within sequence and quality. Usually sequence records will be read sequentially from a file by iteration.</p><pre><code class="language-jlcon">using BioSequences
reader = FASTQ.Reader(open(&quot;hg38.fastq&quot;, &quot;r&quot;))
for record in reader
    # Do something
end
close(reader)</code></pre><p>Reading in a record from a FASTQ formatted file will give you a variable of type <code>FASTQ.Record</code>.</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.FASTQ.Record" href="#BioSequences.FASTQ.Record"><code>BioSequences.FASTQ.Record</code></a> — <span class="docstring-category">Type</span>.</div><div><div><pre><code class="language-none">FASTQ.Record()</code></pre><p>Create an unfilled FASTQ record.</p></div></div><div><div><pre><code class="language-none">FASTQ.Record(data::Vector{UInt8})</code></pre><p>Create a FASTQ record object from <code>data</code>.</p><p>This function verifies and indexes fields for accessors. Note that the ownership of <code>data</code> is transferred to a new record object.</p></div></div><div><div><pre><code class="language-none">FASTQ.Record(str::AbstractString)</code></pre><p>Create a FASTQ record object from <code>str</code>.</p><p>This function verifies and indexes fields for accessors.</p></div></div><div><div><pre><code class="language-none">FASTQ.Record(identifier, sequence, quality; offset=33)</code></pre><p>Create a FASTQ record from <code>identifier</code>, <code>sequence</code> and <code>quality</code>.</p></div></div><div><div><pre><code class="language-none">FASTQ.Record(identifier, description, sequence, quality; offset=33)</code></pre><p>Create a FASTQ record from <code>identifier</code>, <code>description</code>, <code>sequence</code> and <code>quality</code>.</p></div></div></section><p>Various getters and setters are available for <code>FASTQ.Record</code>s:</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.FASTQ.hasidentifier" href="#BioSequences.FASTQ.hasidentifier"><code>BioSequences.FASTQ.hasidentifier</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">hasidentifier(record::Record)</code></pre><p>Checks whether or not the <code>record</code> has an identifier.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.FASTQ.identifier" href="#BioSequences.FASTQ.identifier"><code>BioSequences.FASTQ.identifier</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">identifier(record::Record)::String</code></pre><p>Get the sequence identifier of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.FASTQ.hasdescription" href="#BioSequences.FASTQ.hasdescription"><code>BioSequences.FASTQ.hasdescription</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">hasdescription(record::Record)</code></pre><p>Checks whether or not the <code>record</code> has a description.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.FASTQ.description" href="#BioSequences.FASTQ.description"><code>BioSequences.FASTQ.description</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">description(record::Record)::String</code></pre><p>Get the description of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.FASTQ.hassequence" href="#BioSequences.FASTQ.hassequence"><code>BioSequences.FASTQ.hassequence</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">hassequence(record::Record)</code></pre><p>Checks whether or not a sequence record contains a sequence.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.FASTQ.sequence-Tuple{BioSequences.FASTQ.Record,UnitRange{Int64}}" href="#BioSequences.FASTQ.sequence-Tuple{BioSequences.FASTQ.Record,UnitRange{Int64}}"><code>BioSequences.FASTQ.sequence</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-none">sequence(record::Record, [part::UnitRange{Int}])::BioSequences.DNASequence</code></pre><p>Get the sequence of <code>record</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.FASTQ.hasquality" href="#BioSequences.FASTQ.hasquality"><code>BioSequences.FASTQ.hasquality</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">hasquality(record::Record)</code></pre><p>Check whether the given FASTQ <code>record</code> has a quality string.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.FASTQ.quality" href="#BioSequences.FASTQ.quality"><code>BioSequences.FASTQ.quality</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">quality(record::Record, [offset::Integer=33, [part::UnitRange]])::Vector{UInt8}</code></pre><p>Get the base quality of <code>record</code>.</p></div></div><div><div><pre><code class="language-none">quality(record::Record, encoding_name::Symbol, [part::UnitRange])::Vector{UInt8}</code></pre><p>Get the base quality of <code>record</code> by decoding with <code>encoding_name</code>.</p><p>The <code>encoding_name</code> can be either <code>:sanger</code>, <code>:solexa</code>, <code>:illumina13</code>, <code>:illumina15</code>, or <code>:illumina18</code>.</p></div></div></section><p>To write a <code>BioSequence</code> to FASTQ file, you first have to create a <code>FASTQ.Record</code>:</p><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioSequences.FASTQ.Record-Tuple{AbstractString,Union{Nothing, AbstractString},Any,Array{T,1} where T}" href="#BioSequences.FASTQ.Record-Tuple{AbstractString,Union{Nothing, AbstractString},Any,Array{T,1} where T}"><code>BioSequences.FASTQ.Record</code></a> — <span class="docstring-category">Method</span>.</div><div><div><pre><code class="language-none">FASTQ.Record(identifier, description, sequence, quality; offset=33)</code></pre><p>Create a FASTQ record from <code>identifier</code>, <code>description</code>, <code>sequence</code> and <code>quality</code>.</p></div></div></section><p>As always with julia IO types, remember to close your file readers and writer after you are finished.</p><footer><hr/><a class="previous" href="../fasta/"><span class="direction">Previous</span><span class="title">FASTA formatted files</span></a><a class="next" href="../twobit/"><span class="direction">Next</span><span class="title">2bit formatted files</span></a></footer></article></body></html>
