<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Readme · Kpax3.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL="."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="assets/documenter.js"></script><script src="siteinfo.js"></script><script src="../versions.js"></script><link href="assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>Kpax3.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li class="current"><a class="toctext" href>Readme</a><ul class="internal"><li><a class="toctext" href="#About-1">About</a></li><li><a class="toctext" href="#Reference-1">Reference</a></li><li><a class="toctext" href="#Installation-1">Installation</a></li><li><a class="toctext" href="#Usage-1">Usage</a></li><li><a class="toctext" href="#License-1">License</a></li></ul></li><li><a class="toctext" href="autodocs/">Docstrings</a></li></ul></nav><article id="docs"><header><nav><ul><li><a href>Readme</a></li></ul></nav><hr/><div id="topbar"><span>Readme</span><a class="fa fa-bars" href="#"></a></div></header><h1><a class="nav-anchor" id="Kpax3-Bayesian-bi-clustering-of-categorical-data-1" href="#Kpax3-Bayesian-bi-clustering-of-categorical-data-1">Kpax3 - Bayesian bi-clustering of categorical data</a></h1><p><a href="https://travis-ci.org/albertopessia/Kpax3.jl"><img src="https://travis-ci.org/albertopessia/Kpax3.jl.svg?branch=master" alt="Build Status"/></a> <a href="https://codecov.io/gh/albertopessia/Kpax3.jl"><img src="https://codecov.io/gh/albertopessia/Kpax3.jl/branch/master/graph/badge.svg" alt="Coverage"/></a></p><h2><a class="nav-anchor" id="About-1" href="#About-1">About</a></h2><p>Kpax3 is a <a href="http://julialang.org/">Julia</a> package for inferring the group structure of genetic sequences. In general, any multivariate categorical dataset (such as presence/absence data) can be analyzed by Kpax3. Output consists of a clustering of both the rows (statistical units) and columns (statistical variables) of the provided data matrix. It is an improved version of <a href="https://github.com/albertopessia/kpax2/">K-Pax2</a>, providing an MCMC algorithm for a proper Bayesian approach and a genetic algorithm for MAP estimation.</p><h2><a class="nav-anchor" id="Reference-1" href="#Reference-1">Reference</a></h2><p>To know more about the underlying statistical model, refer to the following publications (the first is the primary citation if you use the package):</p><p>Pessia, A. and Corander, J. (2018). Kpax3: Bayesian bi-clustering of large sequence datasets. <em>Bioinformatics</em>, <strong>34</strong>(12): 2132–2133. doi: <a href="https://doi.org/10.1093/bioinformatics/bty056">10.1093/bioinformatics/bty056</a></p><p>Pessia, A., Grad, Y., Cobey, S., Puranen, J. S., and Corander, J. (2015) K-Pax2: Bayesian identification of cluster-defining amino acid positions in large sequence datasets. <em>Microbial Genomics</em>, <strong>1</strong>(1). doi: <a href="http://doi.org/10.1099/mgen.0.000025">10.1099/mgen.0.000025</a></p><h2><a class="nav-anchor" id="Installation-1" href="#Installation-1">Installation</a></h2><p>Kpax3 can be easily installed from within Julia:</p><ul><li>Enter the Pkg REPL-mode by pressing  <code>]</code> in the Julia REPL</li><li>Issue the command <code>add Kpax3</code></li><li>Press the Backspace key to return to the Julia REPL</li></ul><h2><a class="nav-anchor" id="Usage-1" href="#Usage-1">Usage</a></h2><p>The best way to learn how to use Kpax3 is by following the instructions in the <a href="tutorial/Kpax3_fasta_tutorial.jl">fasta tutorial</a> (for genetic sequences) or in the <a href="tutorial/Kpax3_csv_tutorial.jl">csv tutorial</a> (for general categorical data).</p><p>It is also possible to run Kpax3 directly from the command line by using the script available on <a href="https://gist.github.com/albertopessia/fd9df11fb2bdb158ad91936c4638d6fd">GitHub Gist</a>.</p><h2><a class="nav-anchor" id="License-1" href="#License-1">License</a></h2><p>See <a href="LICENSE.md">LICENSE.md</a></p><footer><hr/><a class="next" href="autodocs/"><span class="direction">Next</span><span class="title">Docstrings</span></a></footer></article></body></html>
