<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>Docstrings · GenomicAnnotations.jl</title><link href="https://cdnjs.cloudflare.com/ajax/libs/normalize/4.2.0/normalize.min.css" rel="stylesheet" type="text/css"/><link href="https://fonts.googleapis.com/css?family=Lato|Roboto+Mono" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/default.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.2.0/require.min.js" data-main="../assets/documenter.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link href="../assets/documenter.css" rel="stylesheet" type="text/css"/></head><body><nav class="toc"><h1>GenomicAnnotations.jl</h1><select id="version-selector" onChange="window.location.href=this.value" style="visibility: hidden"></select><form class="search" id="search-form" action="../search/"><input id="search-query" name="q" type="text" placeholder="Search docs"/></form><ul><li><a class="toctext" href="../">Readme</a></li><li class="current"><a class="toctext" href>Docstrings</a><ul class="internal"></ul></li></ul></nav><article id="docs"><header><nav><ul><li><a href>Docstrings</a></li></ul></nav><hr/><div id="topbar"><span>Docstrings</span><a class="fa fa-bars" href="#"></a></div></header><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicAnnotations.@genes" href="#GenomicAnnotations.@genes"><code>GenomicAnnotations.@genes</code></a> — <span class="docstring-category">Macro</span>.</div><div><div><pre><code class="language-none">@genes(chr, exs...)</code></pre><p>Iterate over and evaluate expressions in <code>exs</code> for all genes in <code>chr.genes</code>, returning genes where all expressions evaluate to <code>true</code>. Any given symbol <code>s</code> in the expression will be substituted for <code>gene.s</code>. The gene itself can be accessed in the expression as <code>gene</code> (see example below).</p><pre><code class="language-julia">julia&gt; chromosome = readgbk(&quot;example.gbk&quot;)
Chromosome &#39;example&#39; (5028 bp) with 6 annotations

julia&gt; @genes(chromosome, :feature == &quot;CDS&quot;) |&gt; length
3

julia&gt; @genes(chromosome, length(gene) &lt; 500)
     CDS             3..206
                     /db_xref=&quot;GI:1293614&quot;
                     /locus_tag=&quot;tag01&quot;
                     /codon_start=&quot;3&quot;
                     /product=&quot;TCP1-beta&quot;
                     /protein_id=&quot;AAA98665.1&quot;</code></pre><p>Accessing properties of the returned list of genes returns a view, which can be altered.</p><pre><code class="language-julia">julia&gt; @genes(chromosome, ismissing(:gene)) |&gt; length
2

julia&gt; @genes(chromosome, ismissing(:gene)).gene .= &quot;Unknown&quot;;

julia&gt; @genes(chromosome, ismissing(:gene)) |&gt; length
0</code></pre></div></div></section><pre><code class="language-none">GenomicAnnotations.AbstractGene</code></pre><pre><code class="language-none">GenomicAnnotations.Chromosome</code></pre><pre><code class="language-none">GenomicAnnotations.Gene</code></pre><pre><code class="language-none">GenomicAnnotations.GeneDataView</code></pre><pre><code class="language-none">GenomicAnnotations.GenomicAnnotations</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicAnnotations.Locus" href="#GenomicAnnotations.Locus"><code>GenomicAnnotations.Locus</code></a> — <span class="docstring-category">Type</span>.</div><div><div><p>Struct for storing information on genomic locations. <code>strand</code> can be &#39;+&#39;, &#39;-&#39;, or &#39;.&#39; when the strand is irrelevant.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicAnnotations.addgene!" href="#GenomicAnnotations.addgene!"><code>GenomicAnnotations.addgene!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">addgene!(chr::Chromosome, feature, locus; kw...)</code></pre><p>Add gene to <code>chr</code>. <code>locus</code> can be a <code>Locus</code>, a UnitRange, or a StepRange.</p></div></div></section><pre><code class="language-none">GenomicAnnotations.appendstring</code></pre><pre><code class="language-none">GenomicAnnotations.eval</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicAnnotations.filterseq" href="#GenomicAnnotations.filterseq"><code>GenomicAnnotations.filterseq</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Parse footer (sequence) portion of a GenBank file, returning a <code>String</code>. When <code>BioSequences</code> is available for 0.7 this will be changed to a <code>DNASequence</code>.</p></div></div></section><pre><code class="language-none">GenomicAnnotations.formatsequence</code></pre><pre><code class="language-none">GenomicAnnotations.genes_helper!</code></pre><pre><code class="language-none">GenomicAnnotations.include</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicAnnotations.iscomplement" href="#GenomicAnnotations.iscomplement"><code>GenomicAnnotations.iscomplement</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">iscomplement(gene::Abstract)</code></pre><p>Return <code>true</code> if <code>gene.locus.compliment == &#39;-&#39;</code>, otherwise return <code>false</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicAnnotations.parsechromosome" href="#GenomicAnnotations.parsechromosome"><code>GenomicAnnotations.parsechromosome</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">parsechromosome(lines)</code></pre><p>Parse and return one chromosome entry, and the line number that it ends at.</p></div></div></section><pre><code class="language-none">GenomicAnnotations.parseheader</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicAnnotations.parseposition" href="#GenomicAnnotations.parseposition"><code>GenomicAnnotations.parseposition</code></a> — <span class="docstring-category">Function</span>.</div><div><div><p>Parse lines encoding genomic position, returning the feature as a <code>String</code>, and an instance of <code>Locus</code>.</p></div></div></section><pre><code class="language-none">GenomicAnnotations.printgbk</code></pre><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicAnnotations.pushproperty!" href="#GenomicAnnotations.pushproperty!"><code>GenomicAnnotations.pushproperty!</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">pushproperty!(gene::AbstractGene, name::Symbol, x::T)</code></pre><p>Add a property to <code>gene</code>, similarly to <code>Base.setproperty!(::gene)</code>, but if the property is not missing in <code>gene</code>, it will be transformed to store a vector instead of overwriting existing data.</p><pre><code class="language-julia">julia&gt; eltype(chr.genedata[:EC_number])
Union{Missing,String}

julia&gt; chr.genes[1].EC_number = &quot;EC:1.2.3.4&quot;
&quot;EC:1.2.3.4&quot;

julia&gt; pushproperty!(chr.genes[1], :EC_number, &quot;EC:4.3.2.1&quot;); chr.genes[1].EC_number
2-element Array{String,1}:
 &quot;EC:1.2.3.4&quot;
 &quot;EC:4.3.2.1&quot;

julia&gt; eltype(chr.genedata[:EC_number])
Union{Missing, Array{String,1}}</code></pre></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicAnnotations.readgbk" href="#GenomicAnnotations.readgbk"><code>GenomicAnnotations.readgbk</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">readgbk(filename)</code></pre><p>Parse GenBank-formatted file <code>filename</code>, returning a <code>Chromosome</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="BioCore.sequence" href="#BioCore.sequence"><code>BioCore.sequence</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">sequence(gene::AbstractGene)</code></pre><p>Return genomic sequence for <code>gene</code>.</p></div></div></section><section class="docstring"><div class="docstring-header"><a class="docstring-binding" id="GenomicAnnotations.vectorise" href="#GenomicAnnotations.vectorise"><code>GenomicAnnotations.vectorise</code></a> — <span class="docstring-category">Function</span>.</div><div><div><pre><code class="language-none">vectorise(A::AbstractArray{Union{Missing, T}, 1}) where T</code></pre><p>Convert an array of type Vector{Union{Missing, T}} to Vector{Union{Missing, Vector{T}}}.</p></div></div></section><footer><hr/><a class="previous" href="../"><span class="direction">Previous</span><span class="title">Readme</span></a></footer></article></body></html>
